---
title: "Find samples both sg and 16s"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
source('~/MSK/work/microbiome_db/SQL/scripts/get_data_from_query_OTU.R')
dbListTables(con) %>% data.frame() %>% filter(grepl("asv",.)) %>% 
  filter(!str_detect(.,'asv'))
library(seqinr)
```

```{r}
sg <- get_data_from_query_OTU(0,'shotgun_lookup_ad')

otu_cts <- get_data_from_query_OTU(0,'counts_ag')

otu_ag 
otu_cts <- get_data_from_query_OTU(0,'otu_ag')
```

```{r}
# use the asv version
# try the CART project samples
# get the asv cts for those samples

pheno <- read_csv('~/melodyP/data/cleaned_pheno/final_pheno.csv') 
samps <- pheno %>%  pull(sampleid)

cts <- get_data_from_query_OTU(0,'asv_counts_ag')
CTS <- cts %>% 
  filter(sampleid %in% samps) %>% 
  select(asv_key, sampleid, count) %>% 
  spread(key = 'sampleid', value = 'count', fill = 0)

CTS %>% 
  write_csv('../data/cart_asv_cts.csv')

seqs <- get_data_from_query_OTU(0,'asv_sequences_ag')

SEQS <- seqs %>% 
  filter(key %in% CTS$asv_key)


write.fasta(sequences = as.list(SEQS$asv_sequence), 
              names = SEQS$key,
              file.out = '../data/cart_asv_seq.fa')

```

