---
title: "Mouse D7 DE analysis"
author: "Anqi Dai"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F)
```

```{r}
library(corncob)
library(tidyverse)
library(vdbR)
library("phyloseq")
connect_database()

```

```{r}
pheno <- readxl::read_excel('../data/05_meta_with_alpha.xlsx') %>% 
  filter(day == 'D7')

# getting the cts from shotgun
fns <- list.files('../data/metaphlan/', pattern = 'profile', full.names = T)
metaphlan <- fns %>% 
  set_names(fns) %>% 
  map(~ read_tsv(., skip = 4) %>% 
          select(species  = `#clade_name`, counts = estimated_number_of_reads_from_the_clade) %>% 
          filter(str_detect(species, 's__'))) %>% 
  bind_rows(.id = 'sampleid') %>% 
  mutate(sampleid = str_extract(sampleid, 'C00....')) %>% 
  filter(sampleid %in% pheno$sampleid) %>% 
  # filter out virus ...
  filter(!str_detect(species, 'k__Viruses'))
```


# Corncob 

```{r}
# it seems the 16s data will work better in this case
# use the 16s data instead

cts <- metaphlan %>% 
  spread('sampleid','counts', fill = 0) %>% 
  column_to_rownames('species')

# I think I need to build a phyloseq object for this
all.equal(pheno$sampleid, colnames(cts))

sampledata <- pheno %>% 
  column_to_rownames('sampleid') %>% 
  mutate(grp = factor(grp, levels = c('BM','BMT')))

samples = sample_data(sampledata)

# assemble a phyloseq object
OTU = otu_table(cts, taxa_are_rows = T)

physeq = phyloseq(OTU,  samples)

```


```{r}
# You may see NAs in the model summary because Wald testing is invalid.
#Likelihood ratio testing can be used, but valid standard errors cannot be calculated.
set.seed(1)
dv_analysis <- differentialTest(
  formula = ~ grp,
  phi.formula = ~ grp,
  formula_null = ~ 1,
  phi.formula_null = ~ grp,
  test = "LRT", boot = FALSE,
  data = physeq,
  fdr_cutoff = 0.25) 
```

# Maaslin2

```{r}
total <- metaphlan %>%
  group_by(sampleid) %>% 
  summarise(total = sum(counts))

relab <- metaphlan %>% 
  left_join(total) %>% 
  mutate(relab = counts/total)
```


```{r}
library(Maaslin2)
cts_relab <- relab %>% 
  select(species, sampleid, relab) %>% 
  spread('species','relab', fill = 0) %>% 
  column_to_rownames('sampleid')

all.equal(rownames(sampledata), rownames(cts_relab))

fit_data <- Maaslin2(
    input_data = cts_relab, 
    input_metadata = sampledata, 
    normalization = "TSS",
    min_abundance = 0.0,
    min_prevalence = 0.0,
    max_significance = 0.25,
    output = "../data/output_shotgun-Maaslin2", 
    fixed_effects = c("grp"),
    reference = c("grp,BM"))

sig <- read_tsv('../data/output_shotgun-Maaslin2/significant_results.tsv') %>% 
  mutate(feature = str_replace_all(feature, '\\.','\\|'))

```

# metagenomeseq

```{r}
library(metagenomeSeq)

pheno_data <- pheno %>% 
  column_to_rownames('sampleid') %>% 
  mutate(grp = factor(grp, levels = c('BM','BMT')))
  
all.equal(colnames(cts) , pheno$sampleid)



# assemble the object for metagenomeSeq

phenotypeData <- AnnotatedDataFrame(pheno_data)

d7 <- newMRexperiment(cts,phenoData=phenotypeData)

p = cumNormStatFast(d7)
```

```{r}
d7 = filterData(d7, present = 2, depth = 1) 
d7 <- cumNorm(d7, p = p) 
pd <- pData(d7) 
mod <- model.matrix(~1 + grp, data = pd) 
d7_res = fitFeatureModel(d7, mod) 


whole <- tibble(species = d7_res@taxa,
                pval = d7_res@pvalues) %>% 
  mutate(padj = p.adjust(pval, method = 'BH'))
q0.25_res <- whole %>% 
  filter(padj < 0.25)

```
# Three tools comparison

```{r}
library(ggvenn)
res_list <- list(metagenomeSeq = q0.25_res$species,
                 MaAsLin2 = sig$feature,
                 Corncob = dv_analysis$significant_taxa)

ggvenn(res_list, show_percentage = F)

# the sig results (FDR < 0.25) from MaAsLin2
sig$feature

# the sig results (FDR < 0.25) from metagenomeSeq
q0.25_res$species

# the sig results (FDR < 0.25) from Corncob
dv_analysis$significant_taxa
```
```{r}

# what is the taxa annotation for the shared 1
all3 <- intersect(intersect(q0.25_res$species,sig$feature), dv_analysis$significant_taxa)
# the intersected sig results between maaslin2 and metagenomicSeq
mas_meta <- intersect(q0.25_res$species,sig$feature)
mas_meta

# the intersected sig results between corncob and metagenomicSeq
intersect(q0.25_res$species,dv_analysis$significant_taxa)

# the intersected sig results between corncob and maaslin2
intersect(sig$feature,dv_analysis$significant_taxa)

# plot the boxplot relab of those asv
cts_relab %>% 
  rownames_to_column('sampleid') %>% 
  gather('species', 'relab', names(.)[2]:names(.)[ncol(.)]) %>% 
  filter(species %in% union(q0.25_res$species,union(sig$feature, dv_analysis$significant_taxa))) %>% 
  left_join(pheno %>% select(sampleid, grp)) %>% 
  mutate(spp = str_extract(species, 's__.+$')) %>% 
  split(.$spp) %>% 
  imap(function(df_, Name_) {
    ggpubr::ggboxplot(data = df_,x = 'grp', y ='relab',  add = 'jitter',
                      title = Name_,
                    add.params = list(color = 'forestgreen', alpha = 0.5, size = 3, shape = 16))
  })
```

