---
title: "figure out the transfer file names"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
```

```{r}
fl <- read_csv('~/pipeline/scripts/dada2_pipeline/data/Florida_pool_set91_samples_list.csv') %>% 
  filter(yes_16s == 'TRUE' & str_detect(Experiment, 'University of Florida'))

folder <- normalizePath('../data/downloaded/')

ret <- fl %>% 
  select(OligoID, sampleid = `Sample ID`, experiment = Experiment) %>% 
  mutate(R1 = str_glue('{folder}/{OligoID}_R1.fastq.gz')) %>% 
  mutate(R2 = str_glue('{folder}/{OligoID}_R2.fastq.gz')) 

att <- ret %>% 
  select(OligoID, R1, R2)

att %>% 
  write_csv('../data/florida_samples.csv')

ret %>% 
  select(R1, R2) %>% 
  gather() %>% 
  mutate(cmd = str_glue('mv {value} /Users/daia1/MSK/work/projects/MSS_pipeline-/scripts/dada2_pipeline/data/downloaded/send')) %>% 
  select(cmd) %>% 
  write_csv('../scripts/mv_files.sh', col_names = F)

ret %>% 
  mutate(R1 = str_replace(R1, '/Users/daia1/MSK/work/projects/MSS_pipeline-/scripts/dada2_pipeline/data/downloaded/','')) %>% 
  mutate(R2 = str_replace(R2, '/Users/daia1/MSK/work/projects/MSS_pipeline-/scripts/dada2_pipeline/data/downloaded/','')) %>% 
  write_csv('../data/Florida_pool_set91_samples_list_match.csv')

```

```{r}
# Oriana neuroblastoma 16s samples
nb <- readxl::read_excel('~/Downloads/Oriana16S_5.17.21.xlsx') %>% 
  filter(`Run.Number` != 'NA') %>% 
  distinct(`Directory.16S`) %>% 
  filter(str_detect(`Directory.16S`, '/Volumes/')) %>%
  mutate(fd = str_replace(`Directory.16S`,'/Volumes/castoricenter/Human.Sequencing.Data/Miseq/','')) %>% 
  mutate(cmd = str_glue('rsync --progress --partial -avz {fd} daia1@lilac.mkscc.org:~/my_workdir/samples/amplicon')) %>% 
  select(cmd)

nb %>% 
  write_csv('../data/upload_O.sh', col_names = F)
```


```{r}
# 2021-10-18
# check the second batch of Florida samples
library(tidyverse)
library(vdbR)
connect_database('~/dbConfig.txt')
get_table_from_database('asv_alpha_diversity_ag')
list_table_from_database('blast')
get_table_from_database('asv_annotation_blast_ag')
could <- asv_alpha_diversity_ag %>% 
  filter(uploaded_date == '2021-09-28' | uploaded_date == '2021-10-04') %>% 
  filter(! str_detect(sampleid, "KM")) %>% 
  filter(str_detect(oligos_id, 'pool1094.1095')) 

samp <- could %>% 
  pull(sampleid)

cts <- get_counts_subset(samp) %>% 
  left_join(asv_annotation_blast_ag %>% 
              select(asv_key:species)) %>% 
  select(-key, -uploaded_date)

div <- could %>% 
  select(sampleid, simpson_reciprocal:shannon)  

div %>% 
  write_csv('../data/Florida_batch2_diversity.csv')

cts %>% 
  write_csv('../data/Florida_batch2_asv_counts_and_blast_annotation.csv')
```

