---
title: "Is there difference in micriobiota composition between white and African American patients during baseline and peri-engraftment time period"
author: "Anqi Dai"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```

```{r}
library(tidyverse)
library(ggpubr)
library(kableExtra)
# define a function for pretty print using Kable
pretty_print <- function(df) {
  df %>% 
    kable() %>% 
    kable_styling(position = 'left', full_width = F)
}
```

## A look at the data and the ethnicity distribution

```{r}
# the ethnicity data on the MSK set that was used in the nejm paper (“frozen set”) here: smb://skimcs.mskcc.org/vandenbrinklab/deep_sequencing/Clinical Annotation/Ethnicity/ethnicity_nejm_frozen_set.csv
# The ethnicity_simple should be used.
pheno <- read_csv('/Volumes/vandenBrinkLab/deep_sequencing/Clinical Annotation/Ethnicity/ethnicity_nejm_frozen_set.csv') %>% 
  select(-X1, -X1_1)

colnames(pheno)
```

### How many patients per each ethnicity group

```{r}
# get the groups percentage
eth_grp <- pheno %>% 
  count(ethnicity_simple) 

eth_grp %>% pretty_print()

nonwhite <- eth_grp %>% 
  mutate(grp = if_else(ethnicity_simple == 'White', 'white', if_else(ethnicity_simple == 'other', 'other', 'nonwhite'))) %>% 
  group_by(grp) %>% 
  summarise(grp_sum = sum(n)) %>% 
  # removing the other cuz don't know their race
  filter(grp != 'other') %>% 
  mutate(grp_perc = round(grp_sum/sum(grp_sum)*100, 2))
```

### White and African American patients and their percentages

```{r}
# if only include white and aa
aa <- eth_grp %>% 
  filter(ethnicity_simple %in% c('Black/African American','White')) %>% 
  mutate(grp_perc = round(n/sum(n)*100, 2))

aa %>% pretty_print()
```

```{r}
# the patients mrn that gonna be used in this analysis: white and nonwhite (excluding the other)
target <- pheno %>% 
  filter(ethnicity_simple != 'other') %>% 
  select(mrn=patient_id, ethnicity=ethnicity_simple)

# how many patients total
nrow(target)
```

## Find the samples for the patients

```{r}
source('~/db_connect_simple.R')
connect_database(config_file = "~/dbConfig.txt")

castori <- get_table_from_database('samples_castori_ag')
allo <- get_table_from_database('patient_allo_ag')
# get the alpha diversity from the db
# the below function will load the unique samples and the table name can't be changed 
get_table_from_database_predefined_filter('asv_alpha_diversity_ag')
```

```{r}
allo %>% 
  count(source)
```

```{r}
# join with all the necessary info
DF <- target %>% 
  left_join(allo %>% 
              mutate(src = if_else(str_detect(source, 'Cord' ), 
                             'cord',
                             if_else(str_detect(source, 'SBA|CD34'), 
                                     'TCD',
                                     'unmodified'))) %>% 
              mutate(intensity = if_else(intensity %in% c('Ablative','ABLATIVE'), 'ablative',
                                         if_else(intensity %in% c('NONABL','Nonablative'), 'nonablative','reduced'))) %>%  
              mutate(intensity = factor(intensity, levels = c('ablative','reduced','nonablative'))) %>% 
              filter(str_detect(indication, fixed('initial', ignore_case = T))) %>%  
              select(mrn, hct, src, intensity), by  = 'mrn') %>% 
  left_join(castori %>% 
              filter(sampletype == 'Stool' | sampletype == 'stool') %>% 
              select(mrn, sampleid, datecollection), by = 'mrn') %>% 
  mutate(drt = datecollection - hct) 


# look at the distribution of the drt
DF %>% 
  gghistogram('drt', bins = 100, color = 'black', fill = 'blue',
              title = 'Distribution of day relative to transplant with all current samples') +
  scale_y_sqrt()
```


```{r}
# add one more filtering on the drt
All <- DF  %>% 
  # filter on the drt < 1000 & drt > -100
  filter(drt < 1000 & drt > -100)
```


```{r}
# look at the dist of drt
All %>% 
  gghistogram('drt', bins = 100, color = 'black', fill = 'blue',
              title = 'Distribution of day relative to transplant with filtering') +
  scale_y_sqrt()
```

```{r}
# join to get the alpha diversity and also the count total to see if need to filter
ALL <- All %>% 
  left_join(asv_alpha_diversity_ag %>% 
              select(sampleid, simpson_reciprocal, shannon, count_total), by = 'sampleid')
```

```{r}
# look at this all table
# there are some NAs we need to deal with that
# ALL %>% 
#   select(-mrn) %>% 
#   head %>% 
```


**We may not have the alpha div and the total counts in the db for a given sample. The reasons may be: **

- the samples are collected but not sequenced
- or couldn't be sequenced because failed to amplify
- or finished sequencing but haven't gone thru the pipeline thus not in the db (new samples)
- nothing in the sequencing file (no read in the file)

```{r}

ALL_final <- ALL %>% 
  filter(!is.na(count_total)) %>% 
  filter(count_total >= 200) %>% 
  # simplify the ethnicity   
  mutate(eth = if_else(ethnicity == 'White', 'white', 'nonwhite'))
```


## alpha diversity

- Are there ethnicity differences in alpha diversity? 
- Can look at first baseline sample (earliest sample per patient collected day -30 to day -6). 
- Can look at the median diversity per patient in the window day 7-21 as the "periengraftment period".

### baseline

```{r}
# in the baseline window: collected day -30 to day -6
df_baseline <- ALL_final %>% 
  filter(drt >= -30 & drt <= -6) 
```

```{r}
# earliest baseline sample
df_baseline_first <- df_baseline %>% 
  group_by(mrn) %>% 
  arrange(drt) %>% 
  slice(1)  

# how many patients
nrow(df_baseline_first)

```

#### White and African American


##### Split by source  

```{r}
df_baseline_first_black <- df_baseline_first %>% 
  ungroup() %>% 
  filter(ethnicity %in% c('Black/African American', 'White')) %>% 
  mutate(ethnicity = if_else(ethnicity == 'Black/African American', 'black', 'white'))

df_baseline_first_black %>% 
  ggboxplot(x = 'ethnicity', y = 'simpson_reciprocal', add = 'jitter', xlab = 'ethnicity',color = 'ethnicity', palette = 'jco',
            title = 'Alpha div simposon reciprocal comparison') +
  stat_compare_means(comparisons = list(c('white', 'black')),
                    label = "p.signif",
                    method = 'wilcox.test',
                    correct=FALSE) +
  facet_wrap(.~ src) 



```


##### Split by intensity 

```{r}
df_baseline_first_black %>% 
  ggboxplot(x = 'ethnicity', y = 'simpson_reciprocal', add = 'jitter', xlab = 'ethnicity',color = 'ethnicity', palette = 'jco',
            title = 'Alpha div simposon reciprocal comparison') +
  stat_compare_means(comparisons = list(c('white', 'black')),
                    label = "p.signif",
                    method = 'wilcox.test',
                    correct=FALSE) +
  facet_wrap(.~ intensity) 


```


### peri-engraftment

```{r}
# peri 

df_peri <- ALL_final %>% 
  filter(drt >= 7 & drt <= 21) 

# median div in the above window
df_peri_med_div <- df_peri %>% 
  group_by(mrn, eth, ethnicity, src, intensity) %>% 
  summarise(med_sr = median(simpson_reciprocal),
            med_shan = median(shannon))

nrow(df_peri_med_div)
```

#### White and African American

##### Split by source 

```{r}
df_peri_med_div_black <- df_peri_med_div %>% 
  ungroup() %>%  
  filter(ethnicity %in% c('Black/African American', 'White')) %>% 
  mutate(ethnicity = if_else(ethnicity == 'Black/African American', 'black', 'white'))

df_peri_med_div_black %>% 
  ggboxplot(x = 'ethnicity', y = 'med_sr', add = 'jitter', xlab = 'ethnicity',color = 'ethnicity', palette = 'jco',
            title = 'Alpha div simposon reciprocal comparison') +
  stat_compare_means(comparisons = list(c('white', 'black')),
                    label = "p.signif",
                    method = 'wilcox.test',
                    correct=FALSE) +
  facet_wrap(.~ src) + scale_y_sqrt()
  


```

##### Split by intensity 

```{r}
df_peri_med_div_black %>% 
  ggboxplot(x = 'ethnicity', y = 'med_sr', add = 'jitter', xlab = 'ethnicity',color = 'ethnicity', palette = 'jco',
            title = 'Alpha div simposon reciprocal comparison') +
  stat_compare_means(comparisons = list(c('white', 'black')),
                    label = "p.signif",
                    method = 'wilcox.test',
                    correct=FALSE) +
  facet_wrap(.~ intensity) + scale_y_sqrt()
  


```


```{r}
# save the two time period tables for use in other scripts
# df_baseline %>% 
#   write_csv('../data/df_baseline.csv')
# 
# df_peri %>% 
#   write_csv('../data/df_peri.csv')
```

## TSNE

```{r}
#What does TSNE do?
#knitr::include_graphics('~/Desktop/screenshots/Screen Shot 2020-04-09 at 3.36.00 PM.png')
```

```{r}
tsnec <- get_table_from_database('tsne_ag')

# get the t1_scaled and t2_scaled from the db and join to the table
time <- bind_rows(
  df_baseline %>% 
    mutate(period = 'baseline'),
  df_peri %>% 
    mutate(period = 'peri')
) %>% 
  left_join(tsnec %>% select(sampleid, t1_scaled, t2_scaled), by = 'sampleid')
  
```


```{r}
# only with white and black
time %>% 
  filter(ethnicity %in% c('Black/African American', 'White')) %>% 
  mutate(ethnicity = if_else(ethnicity == 'Black/African American', 'black', 'white')) %>% 
  filter(!is.na(t1_scaled)) %>% 
  ggscatter(x = 't1_scaled', y = 't2_scaled', color = 'ethnicity', palette = 'jama', alpha = 0.8, 
            title = 'TSNE for two period of time between white and black') +
  facet_wrap(~period)
```

Overlapping between white and black. No separation.

## Adonis test

The statistical test is called Permutational Multivariate Analysis of Variance Using Distance Matrices (PERMANOVA). It is called Adonis here because the function name is that from the vegan package.

**PERMANOVA is an acronym for “permutational multivariate analysis of variance”1. It is best described as a geometric partitioning of multivariate variation in the space of a chosen dissimilarity measure according to a given ANOVA design, with p‐values obtained using appropriate distribution‐free permutation techniques (see Permutation Based Inference; Linear Models: Permutation Methods). The method is semiparametric, motivated by the desire to perform a classical partitioning, as in ANOVA (hence allowing tests and estimation of sizes of main effects, interaction terms, hierarchical structures, random components in mixed models, etc.), while simultaneously retaining important robust statistical properties of rank‐based nonparametric multivariate methods, such as the analysis of similarities (ANOSIM2), namely, (1) the flexibility to base the analysis on a dissimilarity measure of choice (such as Bray–Curtis, Jaccard, etc.) and (2) distribution‐free inferences achieved by permutations, with no assumption of multivariate normality. Thus, PERMANOVA opens the door for formal partitioning of multivariate data in response to complex experimental designs in a wide variety of contexts: there may be more response variables than sampling units, data may be severely non‐normal, zero‐inflated, ordinal or qualitative (e.g., responses to questionnaires, DNA/RNA sequences, allele frequencies, amino acids, or protein data). Although originally motivated by ecological studies, where variables usually consist of counts of abundances (or percentage cover, frequencies, or biomass) for a large number of species, PERMANOVA is now used across many fields, including chemistry, social sciences, agriculture, medicine, genetics, psychology, economics, and more.**

```{r}
df_baseline <- df_baseline %>% 
  filter(ethnicity %in% c('Black/African American', 'White')) %>% 
  mutate(ethnicity = if_else(ethnicity == 'Black/African American', 'black', 'white'))
```


```{r}
get_table_from_database_predefined_filter('asv_counts_ag')
ANNOT <- get_table_from_database('asv_annotation_blast_ag')
```

### baseline

```{r}
# get the counts of those baseline samples 
cts_bl <- asv_counts_ag %>% 
  filter(sampleid %in% df_baseline$sampleid)

cts_bl %>% 
   count(sampleid)
# good we have all the samples counts

cts_baseline <- cts_bl %>% 
  select(asv_key, sampleid, count) %>% 
  spread(key = 'sampleid', value = 'count', fill = 0) %>% 
  arrange(asv_key) %>% 
  column_to_rownames('asv_key') %>% 
  as.matrix()

#dim(cts_baseline)
# cts_baseline %>% 
#   distinct(asv_key, sampleid)
```

```{r}
# get the genus info for the asv_key
annot <- ANNOT %>% 
  filter(asv_key %in% rownames(cts_baseline)) %>% 
  mutate(ordr =  if_else(ordr == '', str_glue('of_class_{class}'), ordr),
         family =  if_else(family == '', str_glue('of_order_{ordr}'), family),
         genus =  if_else(genus == '', str_glue('of_family_{family}'), genus),
         species =  if_else(species == '', str_glue('of_genus_{genus}'), species)) %>% 
  mutate(taxa_genus = str_glue('k__{kingdom}|p__{phylum}|c__{class}|o__{ordr}|f__{family}|g__{genus}'))


cts_baseline_genus_all <- cts_baseline %>% 
  as.data.frame() %>% 
  rownames_to_column('asv_key') %>% 
  full_join(annot %>%  select(asv_key, taxa_genus), by  = 'asv_key') %>% 
  select(-asv_key) %>% 
  gather(key = 'sampleid', value = 'count', names(.)[1]:names(.)[ncol(.) - 1]) %>% 
  group_by(sampleid, taxa_genus) %>% 
  summarise(cnt = sum(count)) %>% 
  # get the total count from the db to calculate the relab
  left_join(asv_counts_ag %>% distinct(sampleid,count_total ), by = 'sampleid') %>% 
  mutate(relab = cnt/count_total) %>% 
  select(sampleid, taxa_genus, relab) 


N <- 40
  
baseline_genus_top <- cts_baseline_genus_all %>% 
  group_by(taxa_genus) %>% 
  summarise(relab_genus = sum(relab)) %>% 
  arrange(desc(relab_genus)) 
  
cts_baseline_genus_topN <- baseline_genus_top$taxa_genus[1:N]
  
cts_baseline_genus_top <- cts_baseline_genus_all %>% 
  filter(taxa_genus %in% cts_baseline_genus_topN) %>% 
  spread(key = 'taxa_genus', value = 'relab') %>% 
  column_to_rownames('sampleid')


#dim(cts_baseline_genus_top)
```

```{r}
# the pheno for the baseline samples 
cts_baseline_pheno <- df_baseline %>% 
  filter(sampleid %in% rownames(cts_baseline_genus_top)) %>% 
  arrange(sampleid)

#all.equal(rownames(cts_baseline_genus_top), cts_baseline_pheno$sampleid)
```

```{r}
# the adonis test
library(vegan)
dist_bl <- vegdist(cts_baseline_genus_top, method = 'bray')
```


```{r}
# for the PCoA
bl_bc <- cmdscale(dist_bl, k = 2) 

bl_bc %>% 
    as.data.frame() %>% 
    rownames_to_column('sampleid') %>% 
    full_join(df_baseline, by = 'sampleid') %>% 
    ggscatter(x = 'V1', y = 'V2', color =  'ethnicity',  palette = 'jco') +
    labs(x = 'PCo 1', y = 'PCo 2', title = 'PCoA (BC distance) between ethnicity') 
```



#### The actual test

```{r}
pmv <- adonis(cts_baseline_genus_top ~ ethnicity, data=cts_baseline_pheno, permutations=999, method = 'bray')
pmv
#densityplot(permustats(pmv))
```

**There is a sig difference between the groups black and white. Possibly due to the big sample size here.**

```{r}
# distance based dispersion test
bd <- betadisper(dist_bl, cts_baseline_pheno$ethnicity)
boxplot(bd)
anova(bd)             
#permutest(bd)
```
 
**The within group dispersion is the same (no sig difference)**

#### The top 20 contributing genera

```{r}
# which genus are responsible for the difference
# The simper functions performs pairwise comparisons of groups of sampling units and finds the contribution of each species to the average between-group Bray-Curtis dissimilarity. Although the method is called simper, it really studied dissimilarities instead of similarities 
sim <- simper(cts_baseline_genus_top , group = cts_baseline_pheno$ethnicity)

genus_contri <- summary(sim) %>% 
  pluck('white_black') %>% 
  rownames_to_column('genus') %>% 
  arrange(desc(average)) %>% 
  head(n = 20)

# average	: Species contribution to average between-group dissimilarity.


genus_contri %>% pretty_print()
```



### peri-engraftment

```{r}
df_peri <- df_peri %>% 
  filter(ethnicity %in% c('Black/African American', 'White')) %>% 
  mutate(ethnicity = if_else(ethnicity == 'Black/African American', 'black', 'white'))

```


```{r}
# get the counts of those baseline samples 
cts_peri <- asv_counts_ag %>% 
  filter(sampleid %in% df_peri$sampleid)

# cts_peri %>% 
#   count(sampleid)
# good we have all the samples counts

cts_peri <- cts_peri %>% 
  select(asv_key, sampleid, count) %>% 
  spread(key = 'sampleid', value = 'count', fill = 0) %>% 
  arrange(asv_key) %>% 
  column_to_rownames('asv_key') %>% 
  as.matrix()

#dim(cts_peri)
# cts_baseline %>% 
#   distinct(asv_key, sampleid)
```

```{r}
# get the genus info for the asv_key
annot <- ANNOT %>% 
  filter(asv_key %in% rownames(cts_peri)) %>% 
  mutate(ordr =  if_else(ordr == '', str_glue('of_class_{class}'), ordr),
         family =  if_else(family == '', str_glue('of_order_{ordr}'), family),
         genus =  if_else(genus == '', str_glue('of_family_{family}'), genus),
         species =  if_else(species == '', str_glue('of_genus_{genus}'), species)) %>% 
  mutate(taxa_genus = str_glue('k__{kingdom}|p__{phylum}|c__{class}|o__{ordr}|f__{family}|g__{genus}'))


N <- 40

cts_peri_genus_all <- cts_peri %>% 
  as.data.frame() %>% 
  rownames_to_column('asv_key') %>% 
  full_join(annot %>%  select(asv_key, taxa_genus), by  = 'asv_key') %>% 
  select(-asv_key) %>% 
  gather(key = 'sampleid', value = 'count', names(.)[1]:names(.)[ncol(.) - 1]) %>% 
  group_by(sampleid, taxa_genus) %>% 
  summarise(cnt = sum(count)) %>% 
  # get the total count from the db to calculate the relab
  left_join(asv_counts_ag %>% distinct(sampleid,count_total ), by = 'sampleid') %>% 
  mutate(relab = cnt/count_total) %>% 
  select(sampleid, taxa_genus, relab) 
 
  
peri_genus_top <- cts_peri_genus_all %>% 
  group_by(taxa_genus) %>% 
  summarise(relab_genus = sum(relab)) %>% 
  arrange(desc(relab_genus)) 
  
cts_peri_genus_topN <- peri_genus_top$taxa_genus[1:N]
  
cts_peri_genus_top <- cts_peri_genus_all %>% 
  filter(taxa_genus %in% cts_peri_genus_topN) %>% 
  spread(key = 'taxa_genus', value = 'relab') %>% 
  column_to_rownames('sampleid')

#dim(cts_peri_genus_top)
```

```{r}
# the pheno 
cts_peri_pheno <- df_peri %>% 
  filter(sampleid %in% rownames(cts_peri_genus_top)) %>% 
  arrange(sampleid)

#all.equal(rownames(cts_peri_genus_top), cts_peri_pheno$sampleid)
```

```{r}
# the adonis test
library(vegan)
dist_peri <- vegdist(cts_peri_genus_top, method = 'bray')
```

```{r}
# for the PCoA
peri_bc <- cmdscale(dist_peri, k = 2) 

peri_bc %>% 
    as.data.frame() %>% 
    rownames_to_column('sampleid') %>% 
    full_join(df_peri, by = 'sampleid') %>% 
    ggscatter(x = 'V1', y = 'V2', color =  'ethnicity',  palette = 'jco') +
    labs(x = 'PCo 1', y = 'PCo 2', title = 'PCoA (BC distance) between ethnicity') 
```

#### The actual test

```{r}
pmv <- adonis(cts_peri_genus_top ~ ethnicity, data=cts_peri_pheno, permutations=999, method = 'bray')
pmv
#densityplot(permustats(pmv))
```

**There is a sig difference between the group black and white**

```{r}
# distance based dispersion test
bd <- betadisper(dist_peri, cts_peri_pheno$ethnicity)
boxplot(bd)
anova(bd)
#permutest(bd)
```
 
**Looks the within group dispersion is sig different already.** 
**significant differences may be caused by different within-group variation (dispersion) instead of different mean values of the groups**


#### The top 20 contributing genera
 
```{r}
# which genus are responsible for the difference
# The simper functions performs pairwise comparisons of groups of sampling units and finds the contribution of each species to the average between-group Bray-Curtis dissimilarity. Although the method is called simper, it really studied dissimilarities instead of similarities 
sim <- simper(cts_peri_genus_top , group = cts_peri_pheno$ethnicity)

genus_contri <- summary(sim) %>% 
  pluck('white_black') %>% 
  rownames_to_column('genus') %>% 
  arrange(desc(average)) %>% 
  head(n = 20)

# average	: Species contribution to average between-group dissimilarity.

genus_contri %>% pretty_print()
```

In summary, we concluded there is no real differences based on ethnicity. The alpha and beta diversity visualization suggested so. As for the adonis test, it could be significant due to the large sample size and the confounding within group dispersion.