---
title: "Microbiome Exploratory LBCL 2022-09-13"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# One-time configuration of the MSK RStudio Package Manager
# install.packages("rstudio.prefs")
# rstudio.prefs::use_rstudio_secondary_repo(
#   MSK_RSPM = "http://rspm.mskcc.org/MSKREPO/latest"
# )
# install.packages("biostatR")
library(biostatR)
 library(mskRvis)  #https://github.mskcc.org/pages/datadojo/mskRvis/index.html

library(tidyverse)
library(vdbR)
library(phyloseq)
library(patchwork)
library(lubridate)
if(packageVersion("vdbR") < "0.1.0") {
  stop("vdbR must be at least version 0.1.0")
}
theme_set(theme_classic())
ymd <- lubridate::ymd
connect_database()

if (!"speedyseq" %in% installed.packages()) remotes::install_github("mikemc/speedyseq")
library(speedyseq)
library(purrr)
library(rlang)
library(forcats)
library(ggpubr)
library(labelled)
library(survival)
library(survminer)
library(cmprsk)
library(hildenp)
library(prodlim)
library(Hmisc)
library(ggalluvial)
library(viridis)

# A function that generate a factor from case_when according to the order conditions are presented
#https://stackoverflow.com/questions/49572416/r-convert-to-factor-with-order-of-levels-same-with-case-when
factored_case_when <- function(...) {
  args <- list2(...)
  rhs <- map(args, f_rhs)
  
  cases <- case_when(
    !!!args
  )
  
  exec(fct_relevel, cases, !!!rhs)
}

load('../data/2022_9_13.rdata')
# day relative to cart and the value for the lab test for the table df_lab
```

# pull and filter clinical data
Can skip to load after data is saves and only run load R.data
```{r, include=FALSE}
# obtain
# Load processed clinical data. #df_lab = labratory data; df_msk - original data after running a data wrangling script; data - original data unprocessed.
#load("~/MSK desktop/Projects/MSKCC/CAR-T/master scripts and data/data/R exports/2022_9_13.Rdata")
# the rdata is after the clinical processing, getting the data from redcap and then cleaning the clinical parameters

#df_lab_raw <- read.csv("~/MSK desktop/Projects/MSKCC/CAR-T/Data hub CAR-T projects orig/dataline/processed dataline file/labs_serial_05-25-2022.csv")


# filtering criteria - clinical population. This includes a heterogeneous group of lymphoma types. For toxicity assessment it's fine to analyze these together. However, for efficacy I would restrict to large-B-cell lymphoma


# define the group of LBCL
lbcl_inclusion <- c("DLBCL NOS", "High-grade B-cell lymphoma with MYC and BCL2 and/or BCL6 rearrangement",
                            "High-grade B-cell lymphoma, NOS", "Primary mediastinal B-cell lymphoma",
                            "T-cell rich DLBCL",
                            "Intravascular large B-cell lymphoma", "EBV-positive DLBCL",
                            "Primary cutaneous DLBCL, leg type","DLBCL associated with chronic inflammation",
                            "ALK positive large B-cell lymphoma",
                            "Plasmablastic lymphoma",
                            "Large B-cell lymphoma arising in HHV8-associated multicentric Castleman Disease",
                            "High-grade B-cell lymphoma with 11q aberrations", "Large B-cell lymphoma with IRF4 rearrangement", "Mantle Cell Lymphoma")
# added the "Mantle Cell Lymphoma" 8 additional samples

# this is gonna be the table that has the finalized patient cohort 
df_msk_filt <-  df_msk %>%
  filter(!is.na(dx)) %>%  # only patients with complete information on disease
  filter(dx.factor %in% lbcl_inclusion) %>%  # took out PCNSL
  mutate(dt_car_t = lubridate:: ymd (dt_car_t))

summary(df_msk$dx.factor)
```
# pull and filter clinical data

```{r}

library(vdbR)
connect_database()
get_table_from_database('asv_alpha_diversity_ag')
get_table_from_database("samples_castori_ag")
# merge clinical data with sample data + filter to day -40 to 60
samples <- samples_castori_ag %>%
  mutate(mrn=str_pad(mrn, pad = "0", width = 8)) %>%
  filter(grepl("stool", sampletype, ignore.case = TRUE)) %>%
  select(mrn, sampleid, datecollection, consistency, sampletype) %>%
  inner_join(df_msk_filt,  by=c("mrn"))%>%
#  inner_join(df_msk_filt %>% select(mrn,record_id, dx.factor, dt_car_t,ind_aph_cart_d, crs.factor, objres_100,everCR_100, icans.factor, dur_neut_d, bestres_3lvl_100, everCR_100), by=c("mrn"))%>%
  mutate(dr_car = as.numeric(ymd(datecollection) - ymd(dt_car_t))) %>%
  filter(dr_car > -40 & dr_car <= 100)
```


```{r}
# figure out what are the most recently sequenced samples in the pool1157
new <- asv_alpha_diversity_ag %>% 
  filter(str_detect(oligos_id, 'pool1157'))

length(intersect(samples$sampleid, new$sampleid))

newpt <- new %>% left_join(samples_castori_ag %>% select(mrn, sampleid))

# need to update the samples castori ag table

```


```{r}
## Creating a Phyloseq Object
# Here we create a phyloseq object.  After creation we remove low-depth samples (arbitrarily set to 5k reads, dropping 3 samples)

phy_raw <- vdb_make_phylo(samples, sampleid_col = "sampleid")
depths <- sample_sums(phy_raw)
sum(depths < 5000)

data.frame(lib_size = sample_sums(phy_raw)) %>% rownames_to_column("sampleid") %>%
  ggplot(aes(x=lib_size)) + geom_histogram(bins=30)

summary(sample_sums(phy_raw))
phy_raw <- subset_samples(phy_raw, sample_sums(phy_raw) > 5000)

#pull lab data

# merge labs with filtered dataset (if you want to get creatinine: df_lab %>% filter(lab=="creatinine"))
df_lab_sample <- samples %>%
  select(record_id) %>%
  group_by(record_id) %>%
  slice(1) %>%
  ungroup() %>%
  left_join(df_lab %>%
              select(record_id,performed_dte, lab, result_value, unit_measure:lower_limit, day_rel_car, day_rel_ld ), by ="record_id") %>%
  mutate(result_value = as.numeric(result_value)) %>%
  set_variable_labels(
    performed_dte = "Lab date",
    result_value = "Value",
    unit_measure = "Unit",
    day_rel_car = "Day relative to CAR-T",
    day_rel_ld = "Day relative to lymphodepletion") %>%
  filter(between(day_rel_car,-20, 365))   # filter time of car-t

# # QC labs
# df_lab %>%
#   mutate(result_value = as.numeric(result_value)) %>%
#   mutate(lab = factor(lab),
#          lab = fct_infreq(lab)) %>%
#   group_by (lab) %>%
#   summarise(n=n(), median = median (result_value), min= min(result_value),
#             max = max(result_value),  iqr_1= quantile(result_value, 0.25, na.rm=T), iqr_3= quantile(result_value, 0.75, na.rm=T)) %>%
#   View()

get_table_from_database("asv_annotation_blast_color_ag")

# We filter to remove rare taxa (those undetectable in the lowest-depth sample, so 2 divided by the lowest-depth sample as DADA2 does not consider singletons)
min_abund <- round(2/min(sample_sums(phy_raw)), 4) * .1
phy_filt  = transform_sample_counts(phy_raw, function(x) x / sum(x) )
phy_filt = filter_taxa(phy_filt, function(x) mean(x) > min_abund, TRUE)
sample_sums(phy_filt)
phy = tax_glom(phy_filt, taxrank = "genus")
phy


#save.image("~/MSK desktop/Projects/MSKCC/CAR-T/microbiome mskcc/data/2022_09_13_angel_export/2022_09_13_mb_ad_export.Rdata")

#load("~/MSK desktop/Projects/MSKCC/CAR-T/CAR-T-data-repository/data/2022_6_14_microbiome_data.Rdata")


```



 ```{r gt theme, include=FALSE}
# my_theme <- list(
#   "pkgwide-fn:pvalue_fun" = function(x) gtsummary::style_pvalue(x, digits = 2),
#   "pkgwide-fn:prependpvalue_fun" = function(x) gtsummary::style_pvalue(x, digits = 2, prepend_p = TRUE),
#   "tbl_summary-fn:percent_fun" = function(x) paste0(round(x*100)),
#   "tbl_summary-str:categorical_stat" = "{n} ({p}%)",
#   "tbl_summary-str:continuous_stat" = "{median} [{min} \U2013 {max}]",
#  # "tbl_summary-str:continuous_stat" = "{median} ({p25} - {p75})",
# 
#   "pkgwide-str:print_engine" = "flextable"
# )
# set_gtsummary_theme(my_theme)
 ```

# Sample Exploration
```{r}

mrns_missing <- df_msk_filt[!df_msk_filt$mrn %in% samples$mrn, "mrn"]

print(paste0("Found sequencing data for ", nrow(df_msk_filt)-nrow(mrns_missing), "/", nrow(df_msk_filt), " patients"))
print(paste0("No sequencing data for ", nrow(mrns_missing), "/", nrow(df_msk_filt), " patients"))

samples %>%
  ggplot(aes(dr_car)) +
  geom_histogram(binwidth = 2) +
  scale_x_continuous(breaks=seq(-40, 101, 10)) +
  labs(
  subtitle = "Sample distribution",
    x = 'Days from CAR-T',
    y = 'Sample Count',
    caption = paste0("Found sequencing data for ", nrow(df_msk_filt)-nrow(mrns_missing), "/", nrow(df_msk_filt), " patients\n",
    "No sequencing data for ", nrow(mrns_missing), "/", nrow(df_msk_filt), " patients"))+
  scale_color_manual(values = msk_palette("main")) +
  theme_msk() +
  theme (legend.position = "top")+ 
  set_msk_aes()



samples %>%
filter(!is.na(crs_2_4)) %>% 
  ggplot(aes(dr_car, fill = crs_2_4)) +
  geom_histogram(binwidth = 2, alpha = 0.4, position = "identity") +
  scale_x_continuous(breaks=seq(-40, 101, 10)) +
  labs(  subtitle = "Sample distribution",
    x = 'Days from CAR-T',
    y = 'Sample Count')+
 scale_fill_manual(values = msk_palette("contrast")) +
  theme_msk() +
  set_msk_aes()


samples %>%
filter(!is.na(everCR_100)) %>% 
  ggplot(aes(dr_car, fill = everCR_100)) +
  geom_histogram(binwidth = 2, alpha = 0.4, position = "identity") +
  scale_x_continuous(breaks=seq(-40, 101, 10)) +
  labs(  subtitle = "Sample distribution",
    x = 'Days from CAR-T',
    y = 'Sample Count')+
 scale_fill_manual(values = msk_palette("contrast")) +
  theme_msk() +
  set_msk_aes()

# Generate a table with cutpoints for grouping samples  based on the distribution seen in the histogram

table_samples <- samples %>% 
  mutate(range =cut (dr_car, breaks = c(-35,-10,0, 10, 20, 30,100))) %>% 
  select(range, dr_car) %>% 
    tbl_summary(by =range,
                 type = all_continuous() ~ "continuous2",
    statistic = all_continuous() ~ c("{N_nonmiss}",
                                     "{median} ({p25}, {p75})",
                                      "{min}, {max}")) %>% 
                    modify_header(stat_by =  "{level}") %>% 
  modify_header(label = "") %>% 
  modify_spanning_header(all_stat_cols ()~"**Time windows (days)**")

table_samples$table_body[1,5] <- ""
table_samples

# code sample windows

samples <- samples %>% 
  mutate( range = factored_case_when(   
           dr_car <= -0 ~ "pre-car",
           dr_car <= 10 ~ "1-10",
           dr_car <= 20 ~ "11-20",
           dr_car <= 50 ~ "21-50"))


table_grouped_samples <- samples %>% 
  select(range, dr_car) %>% 
    tbl_summary(by =range,
                 type = all_continuous() ~ "continuous2",
    statistic = all_continuous() ~ c("{N_nonmiss}",
                                     "{median} ({p25}, {p75})",
                                      "{min}, {max}")) %>% 
                    modify_header(stat_by =  "{level}") %>% 
  modify_header(label = "") %>% 
  modify_spanning_header(all_stat_cols ()~"**Time windows (days)**")

table_grouped_samples$table_body[1,5] <- ""
table_grouped_samples
```

# Clinical description
## Patient charactersitics - Table 1


```{r}
# drop unused levels
samples$car_t_product.factor <- droplevels(samples$car_t_product.factor)
samples$dx_lbcl_who <- droplevels(samples$dx_lbcl_who)
samples$lymphodepletion.factor <- droplevels(samples$lymphodepletion.factor)


# dropping levels removes labels, so I reintroduce them here
samples <- samples %>% 
  set_variable_labels(
    dx_lbcl_who = "NHL WHO diagnosis",
    car_t_product.factor = "CAR-T product",
    dt_car_t = "CAR-T date",
    aph_to_cart_d = "Days from aphresis to CAR",
    race.factor = "Race",
    lymphodepletion.factor = "Lymphodepletion")



# generate a dataframe where each patient is a single row
samples_sliced <- samples %>% 
  group_by(record_id) %>% 
  slice(1) %>% 
  ungroup ()

samples_sliced %>% 
  select(calc_age_car, sex.factor, race.factor,
         kps90,     pre_ld_bmi_cat,
         dx_lbcl_who, trans_nhl.factor,  coo_gen,
         dexl_gen, dblhit_any,
        primary_ref.factor, num_rx_line_3lev, auto_hx.factor, allo_hx.factor,
         dt_car_t, dx_to_cart_m, aph_to_cart_d,  bridge_modality, bridge_to_cart,
         car_t_product.factor, lymphodepletion.factor, ds_res_infusion) %>% 
  tbl_summary(  statistic = all_continuous() ~ c("{median} ({min}, {max})")) %>%  
  add_variable_grouping("Bridging" = c("bridge_modality", "bridge_to_cart")
                        ) %>% 
  bold_labels() %>% 
  modify_caption("Baseline Patient Characteristics")
```

## Available labs
```{r}

labs_present <- c("wbc", "alc", "anc", "hb",   "plt",
                  "creatinine", "albumin", "alt", "ast", "alk","tbr", 
                  "ldh", "crp", "ddimer", "fibrinogen", "ferritin", "il1b", "tnfa", "il10", "il6")
p_lab <- df_lab %>% 
  filter(day_rel_car>-10, day_rel_car< 30) %>% 
  filter(!is.na(result_value)) %>% 
  filter (lab %in% labs_present) %>%
  mutate(lab = factor (lab, levels = labs_present)) %>% 
  ggplot(aes (x = day_rel_car , y = result_value))+
  geom_point(alpha = 0.1, size = 1) + 
  geom_smooth(color = "#83276B") +
  facet_wrap(~lab, scales = "free") +
  labs (x = "Days from CAR-T", y = "Value")+
  theme_msk() +
  theme(strip.text.x = element_text(size = 13, face = "bold"))


#pdf("p_lab.pdf", width = 10, height = 10)
p_lab
#dev.off()


```



## Response 
```{r}


samples_sliced %>% 
  select(car_t_product.factor, response_28_3lvl, bestres_3lvl_100) %>% 
  tbl_summary(by = "car_t_product.factor") %>% 
  bold_labels() %>% 
  add_overall() %>% 
  modify_caption("Best Response (day 28 and 100)")

```
## Survival-related outcomes

### Median Follow-up 
```{r, warning=FALSE, message=FALSE}
q <- quantile(prodlim(Hist(tt_os_d/30.42, ev_os)~1, reverse=T, samples_sliced))$quantiles.survival
paste(round(q$quantile[3], 2), " months, IQR: (", round(q$quantile[4],2), "-", round(q$quantile[2],2), ")", sep="")
```

### Overall Survival 

Events: Death    
Censored: Last Follow-up

```{r}
os = survfit(Surv(tt_os_d/30.42, ev_os)~1, data = samples_sliced)


os_plot <-ggsurvplot(os, data =samples_sliced ,
                     risk.table = TRUE, tables.height = 0.15,
                     tables.theme = theme_cleantable(),
                     censor=FALSE,
                     conf.int = T,
                     legend = c("none"),
                     break.x.by=3,
                     xlim = c(0,37),
                     axes.offset=FALSE,
                     xlab="Time in Months",
                     ylab = "OS")

#pdf("OS_entire_cohort.pdf")
os_plot
#dev.off()

tbl_survfit(
  os,
  times = c(12, 24),
  label = "{time} Months") %>% 
  modify_caption("Overall Survival Rates")

  
tbl_survfit(
  os,
  probs = c(0.5),
  header_estimate = "**Months**"
) %>% 
  modify_caption("Median Overall Survival")
```



## Progression Free Survival 

Events: Progression, Relapse, Death  
Censored: Next therapy, last follow-up  

```{r}
pfs = survfit(Surv(tt_pfs_d/30.42, ev_pfs)~1, data = samples_sliced)


pfs_plot <-ggsurvplot(pfs, data =samples_sliced ,
                      risk.table = TRUE, tables.height = 0.15,
                      tables.theme = theme_cleantable(),
                      censor=T,
                      conf.int = F,
                      legend = c("none"),
                      break.x.by=3,
                      xlim = c(0,37),
                      axes.offset=FALSE,
                      xlab="Time in Months",
                      ylab ="PFS")


#pdf("pfs_entire_cohort.pdf")
pfs_plot
#dev.off()

tbl_survfit(
  pfs,
  times = c(12, 24),
  label = "{time} Months"
) %>% 
  modify_caption("PFS Rates")


tbl_survfit(
  pfs,
  probs = c(0.5),
  header_estimate = "**Months**"
) %>% 
  modify_caption("Median PFS")
```


## Event Free Survival 

Events: Progression, Death, new anticancer therapy
Censor: Last follow-up  

```{r}
efs = survfit(Surv(tt_efs_d/30.42, ev_efs)~1, data = samples_sliced)


efs_plot <-ggsurvplot(efs, data =samples_sliced ,
                      risk.table = TRUE, tables.height = 0.15,
                      tables.theme = theme_cleantable(),
                      censor=T,
                      conf.int = F,
                      legend = c("none"),
                      break.x.by=3,
                      xlim = c(0,37),
                      axes.offset=FALSE,
                      xlab="Time in Months",
                      ylab = "EFS")


#pdf("efs_entire_cohort.pdf")
efs_plot
#dev.off()

tbl_survfit(
  efs,
  times = c(12, 24),
  label = "{time} Months"
) %>% 
  modify_caption("EFS Rates")

tbl_survfit(
  efs,
  probs = c(0.5),
  header_estimate = "**Months**"
) %>% 
  modify_caption("Median EFS")

```

### Relapse / Progression Cumulative Incidence  

Events: Relapse, Progression
Competing Risk: Death without Relapse/Progression
Censored: New therapy, last follow-up

```{r}
m<-cuminc(samples_sliced$tt_pfs_d/30.42, samples_sliced$cev_progression)

df_plot<-extract_cuminc(m)
df_plot<-df_plot[which(df_plot$event==1),]

plot=
  ggplot()+
  geom_step(data=df_plot, aes(x=time, y=prob), size = 1.05)+
  theme_bw()+
  theme(legend.position = "none", panel.grid = element_blank(),   legend.title = element_blank(), 
        axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank())+
  labs(y=c("Cumulative Incidence of Relapse"), x=c("Months"))+
  coord_cartesian(ylim=c(0,1),xlim=c(0,37.5))+
  scale_x_continuous(breaks=c(0,3,6,9,12,15,18,21,24,27,30,33,36)) +
  scale_y_continuous(breaks=c(0, 0.2, 0.4, 0.6, 0.8,1))+
  theme(plot.title = element_text(hjust=0.5), text = element_text(size=10)) 


plot +
  theme_msk()
# 
# fit <- survfit(
#   Surv(tt_pfs_d, ifelse(cev_progression != 2, 1, 0)) ~ 1, 
#   data = samples_sliced
# )
# 
# 
# num <- ggsurvplot(
#   fit = fit, 
#   xlim=c(0,36),
#   break.x.by=c(3),
#   risk.table = TRUE,
#   risk.table.y.text = TRUE,
#   ylab = "Months",
#   tables.theme = theme_cleantable(),
#   risk.table.fontsize = 3.2
# )
# plot
# #pdf("relapse_entire_cohort.pdf")
# cowplot::plot_grid(
#   plot, 
#   num$table, 
#   nrow = 2, 
#   rel_heights = c(4, 1), 
#   align = "v", 
#   axis = "b"
# )
#dev.off()

```

```{r, results='hide'}
vals = cuminc_ci(samples_sliced$tt_pfs_d/30.42, samples_sliced$cev_progression, times = c(12, 24))

```


1-year CI of Relapse: `r vals$time1$estimate[1]`, 95% CI: (`r vals$time1$lower.95[1]` , `r vals$time1$upper.95[1]`) 
2-year CI of Relapse: `r vals$time2$estimate[1]`, 95% CI: (`r vals$time2$lower.95[1]` , `r vals$time2$upper.95[1]`) 


### Toxicities
```{r}
samples_sliced %>%
  select(car_t_product.factor,car_t_site.factor, 
         crs_max.factor,crs_2_4, crs_3_4, tt_car_crs_d, tt_crs_res_d, crs_treated,
         icans_max.factor, icans_2_4, icans_3_4, tt_icans_crs_d, tt_icans_crs_d, icans_treated,
         anak_ppx, toci_trt, steroids_trt, tox_icu.factor, ev_bsi,
         death_30, tox_sig, tox_neutropenia.factor,
         dur_neut_d, tox_thrombopenia.factor, dur_thrombocyt_d) %>% 
  tbl_summary(
              type = c(tt_car_crs_d, tt_crs_res_d, tt_icans_crs_d, tt_icans_crs_d, dur_neut_d,dur_thrombocyt_d ) ~ "continuous") %>% 
  bold_labels() %>% 
  modify_caption("Toxicities") 

  
samples_sliced %>% 
  filter(!is.na(crs_max.factor)) %>% 
  select(crs_max.factor) %>% 
  group_by(crs_max.factor) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(crs_max.factor = fct_rev(crs_max.factor)) %>% 
  ggplot()+
  aes(1, n, fill = crs_max.factor) + 
           geom_bar( stat="identity") +
  scale_fill_manual(values = rev(c("#B6B6B1", "#8FC7E8", "#007CBA","#F29934"))) +
   scale_x_discrete("CRS")+
  theme_msk() +
  set_msk_aes()+
  theme(legend.position = "none") +
samples_sliced %>% 
  filter(!is.na(icans_max.factor)) %>% 
  select(icans_max.factor) %>% 
  group_by(icans_max.factor) %>% 
  count() %>% 
  ungroup() %>% 
  mutate(icans_max.factor = fct_rev(icans_max.factor)) %>% 
  ggplot()+
  aes(1, n, fill = icans_max.factor) + 
           geom_bar( stat="identity") +
  scale_fill_manual(values = rev(c("#B6B6B1", "#8FC7E8", "#007CBA","#F29934","#E92076"))) +
   scale_x_discrete("ICANS")+
  scale_y_continuous("")+
    theme_msk() +
  set_msk_aes() +
    theme(axis.title.x = element_text(),
        axis.text.y = element_blank(),
        legend.position = "right") +
  labs(fill = "Grade")

```
#Microbiome analysis
```{r}
# alpha_df <- sample_data(phy_raw) %>% data.frame() %>% 
#   rownames_to_column("Sample") %>%
#   left_join(estimate_richness(phy_raw, measures = "InvSimpson") %>% 
#               rownames_to_column("Sample"))
# ggplot(alpha_df, aes(x=everCR_100, y=InvSimpson, color =)) + geom_boxplot(outlier.colour = NA, alpha = 0.5) + geom_jitter(height = 0) +
# ggplot(alpha_df, aes(x=crs.factor, y=InvSimpson)) + geom_boxplot(outlier.colour = NA) + geom_jitter(height = 0) +
# ggplot(alpha_df, aes(x=objres_100, y=InvSimpson)) + geom_boxplot(outlier.colour = NA) + geom_jitter(height = 0) 
```

The resulting filtered phyloseq object is then ordinated with PCoA based on Bray-curtis distances:
```{r}
ord <- ordinate(phy, "PCoA", "bray")

plot_df <- samples %>%
  left_join(ord$vectors %>% as.data.frame() %>% select(1:5) %>% rownames_to_column("sampleid")) %>%
  left_join(estimate_richness(phy_raw, measures = "InvSimpson")%>% rownames_to_column("sampleid"))




ggplot(plot_df, aes(x=Axis.1, y=Axis.2)) + geom_point(color="black",stroke=NA,   shape=21, aes(fill=dr_car)) + scale_fill_viridis() +
  labs(fill = "Days") +  stat_ellipse(aes(group=range, color=range), size =1.3, linetype = "dashed") +theme_msk() +
  guides(color = guide_legend( "",nrow = 2, ))+ 
    ggtitle("Days form CAR") +

  
ggplot(plot_df, aes(x=Axis.1, y=Axis.2)) + geom_point(color="black",stroke=NA,   shape=21, aes(fill=log10(InvSimpson))) + scale_fill_gradient2() +theme_msk() +
  ggtitle("Inv. Simpson") +
    
ggplot(plot_df, aes(x=Axis.1, y=Axis.2)) + geom_point(color="black",stroke=NA,   shape=21, aes(fill=age_car)) + scale_fill_gradient() +theme_msk() +
  ggtitle("Age") 



plot_ordination(phy, ord,  color="dur_neut_d", title="dur_neut_d")  + scale_color_continuous(trans = "log10") + scale_color_viridis_c() + theme_msk() 


plot_ordination(phy, ord,  color="crs_2_4", title="crs_2_4") +stat_ellipse() + scale_color_manual(values = msk_palette("main"))+ theme_msk() +
  set_msk_aes()+

plot_ordination(phy, ord,  color="icans_2_4", title="icans_2_4") +stat_ellipse() + scale_color_manual(values = msk_palette("main"))+ theme_msk() +
  set_msk_aes() +


plot_ordination(phy, ord,  color="tox_sig", title="tox_sig") +stat_ellipse() + scale_color_manual(values = msk_palette("main"))+ theme_msk() +
  set_msk_aes()


plot_ordination(phy, ord,  color="everCR_100", title="Day 100 CR") +stat_ellipse() + scale_color_manual(values = msk_palette("main"))+ theme_msk() +
  set_msk_aes()+

plot_ordination(phy, ord,  color="bestres_3lvl_100", title="Day 100 Best Response") + stat_ellipse() +   scale_color_manual(values = c("#007CBA", "#40B4E5", "#E92076","#F29934","#E92076")) + theme_msk() 




ggplot(plot_df, aes(x=Axis.1, y=Axis.2, color=interaction(crs.factor,icans.factor ))) + geom_point() + stat_ellipse()+ scale_color_manual(values = c("#007CBA", "#40B4E5", "#E92076","#F29934","#E92076")) + theme_msk() 
ggplot(plot_df, aes(x=Axis.1, y=Axis.2, color=objres_100)) + geom_point() + stat_ellipse() +  scale_color_manual(values = msk_palette("main"))+ theme_msk()

ggplot(plot_df, aes(x=Axis.1, y=Axis.2)) + geom_point(color="black",stroke=NA,   shape=21, aes(fill=log10(InvSimpson))) +
  stat_ellipse(aes(group=objres_100, color=objres_100)) + scale_color_manual(values = msk_palette("main"))+ theme_msk()



ggplot(plot_df, aes(x=Axis.1, y=Axis.2)) + geom_point(color="black",stroke=NA,   shape=21, aes(fill=log10(InvSimpson))) +
  stat_ellipse(aes(group=everCR_100, color=everCR_100)) + scale_color_manual(values = msk_palette("main"))+ theme_msk()


ggplot(plot_df, aes(x=Axis.1, y=Axis.2, color=range)) + geom_point() + stat_ellipse()  + scale_color_manual(values = msk_cols(c("msk_blue", "light_blue", "light_green","cb_violet1"))) + theme_msk()


ggplot(plot_df %>% filter(dr_car < 60), aes(x=Axis.1, y=Axis.2, color=dr_car)) + geom_point(size=3, alpha=.8) + scale_color_viridis_c()  +theme_msk()

# ggplot(plot_df %>% filter(dr_car < 30), aes(x=Axis.1, y=Axis.2, color=dr_car)) + geom_point(size=3, alpha=.8) +  scale_color_gradientn(colors = msk_palette("blues", type = "continuous"))  +theme_msk()


# 
# ggplot(plot_df %>% filter(!is.na(bestres_3lvl_100)) %>% filter(dr_car < 30), aes(x=Axis.1, y=Axis.2, color=bestres_3lvl_100)) + geom_point(size=3, alpha=.8)+theme_classic()  + stat_ellipse()
#ggplot(plot_df %>% filter(!is.na(tox_sig)) %>% filter(dr_car <0), aes(x=Axis.1, y=Axis.2, color=tox_sig)) + geom_point(size=3, alpha=.8)+theme_classic()  + stat_ellipse()

# ggplot(plot_df, aes(x=dr_car, y=InvSimpson, colggplot(plot_df, aes(x=dr_car, y=InvSimpson, color=interaction(crs.factor,icans.factor)) + geom_point(size=3, alpha=.8)+theme_classic()  +geom_smooth()
# ggplot(plot_df %>% filter(!is.na(crs.factor)), aes(x=dr_car, y=InvSimpson)) + geom_point(size=3, alpha=.8) +theme_classic()  +geom_smooth()
# ggplot(plot_df %>% filter(!is.na(crs.factor)), aes(x=dr_car, y=InvSimpson, color=crs.factor)) + geom_point(size=3, alpha=.8) +theme_classic()  +geom_smooth()
# ggplot(plot_df %>% filter(!is.na(icans.factor)), aes(x=dr_car, y=InvSimpson, color=icans.factor)) + geom_point(size=3, alpha=.8) +theme_classic()  +geom_smooth()
# ggplot(plot_df %>% filter(!is.na(everCR_100)), aes(x=dr_car, y=InvSimpson, color=everCR_100)) + geom_point(size=3, alpha=.8) +theme_classic()  +geom_smooth()

plot_df %>% 
  group_by(mrn, everCR_100) %>% 
         filter(dr_car > 0 & dr_car < 30) %>% 
   summarise(med_alpha_tox = median(InvSimpson, na.rm = T)) %>% 
  filter(!is.na(everCR_100)) %>% 
  ggplot()+
  aes(x=everCR_100, y=med_alpha_tox) +
  geom_boxplot(outlier.colour = NA, alpha =0.5)+ 
  geom_jitter(size=1, alpha=.8, height = 0, width = .1)  + scale_fill_manual(values = msk_palette("main"))+theme_msk() 



plot_bar(phy,fill="class") + facet_wrap(~objres_100, scales = "free_x")+theme_msk() 

  # or
asv_blast_color_genus <- asv_annotation_blast_color_ag %>% select(genus, color, color_base, color_label_group, color_label_group_distinct) %>% distinct()

these_colors <- asv_blast_color_genus$color
names(these_colors) <- asv_blast_color_genus$color_label_group_distinct
ggplot( psmelt(physeq = phy) %>% 
          left_join(asv_blast_color_genus, by="genus"),
                        aes(x=reorder(paste(Sample, dr_car), dr_car), 
                            y=Abundance,
                            fill=color_label_group_distinct) ) +
  geom_bar(stat = "identity",position="fill",width = 1) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90),
        axis.text.y = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = these_colors, breaks = names(these_colors), guide="none") +
  facet_grid(~everCR_100, scales = "free", space="free") +
  scale_y_continuous(expand=c(0,0)) +
  scale_x_discrete(expand=c(0,0)) +
  labs(x="Group", y="Relative Abundance") +
  theme(axis.text=element_text(size=10, color = "black", hjust = 1)) +
  theme(strip.text.x = element_text(size = 10))+
  theme_msk() 

ggplot( psmelt(physeq = phy) %>% 
          left_join(asv_blast_color_genus, by="genus"),
                        aes(x=reorder(paste(Sample, dr_car), dr_car), 
                            y=Abundance,
                            fill=color_label_group_distinct) ) +
  geom_bar(stat = "identity",position="fill",width = 1) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90),
        axis.text.y = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = these_colors, breaks = names(these_colors), guide="none") +
  facet_grid(~everCR_100, scales = "free", space="free") +
  scale_y_continuous(expand=c(0,0)) +
  scale_x_discrete(expand=c(0,0)) +
  labs(x="Group", y="Relative Abundance") +
  theme(axis.text=element_text(size=10, color = "black", hjust = 1)) +
  theme(strip.text.x = element_text(size = 10))+
  theme_msk() 

ggplot( psmelt(physeq = phy) %>% 
          left_join(asv_blast_color_genus, by="genus"),
                        aes(x=reorder(paste(Sample, dr_car), dr_car), 
                            y=Abundance,
                            fill=color_label_group_distinct) ) +
  geom_bar(stat = "identity",position="fill",width = 1) +
  theme_classic() +
  theme(axis.text.x = element_text(angle=90),
        axis.text.y = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = these_colors, breaks = names(these_colors), guide="none") +
  facet_grid(~crs_2_4, scales = "free", space="free") +
  scale_y_continuous(expand=c(0,0)) +
  scale_x_discrete(expand=c(0,0)) +
  labs(x="Group", y="Relative Abundance") +
  theme(axis.text=element_text(size=10, color = "black", hjust = 1)) +
  theme(strip.text.x = element_text(size = 10))+
  theme_msk() 
```


## Maaslin Modeling
Skip chunck and run this separately because it causes  an error in RMD
```{r eval = FALSE, echo = FALSE}

library(Maaslin2)

masdat <- left_join(
    tax_table(phy) %>% as.data.frame()  %>% rownames_to_column("key"),
   t(otu_table(phy)) %>% as.data.frame() %>% rownames_to_column("key")
) %>% pivot_longer(cols = -c(key:species)) %>% 
  select(-c(key, kingdom, phylum,class,    ordr,  family,   species)) %>% 
  pivot_wider(names_from="genus", values_fn = sum, values_fill = 0) %>% 
  column_to_rownames("name")
# 
# res<- Maaslin2(
#     masdat, 
#     sample_data(phy)%>% data.frame() %>% filter(!is.na(objres_100)) %>% 
#     filter(dr_car <30 & dr_car > -30 ),
#     'maaslin_objres100t',
#     fixed_effects = c('objres_100' ),
#     random_effects = c('mrn'),
#     standardize = FALSE, plot_heatmap = TRUE)
# res_precar<- Maaslin2(
#     masdat, 
#     sample_data(phy)%>% data.frame() %>% filter(!is.na(objres_100)) %>% 
#     filter(dr_car < 0  ),
#     'maaslin_objres100t_precar',
#     fixed_effects = c('objres_100' ),
#     random_effects = c('mrn'),
#     standardize = FALSE, plot_heatmap = TRUE)
# 




res_cr100_1_10<- Maaslin2(
    masdat, 
    sample_data(phy)%>% data.frame() %>% filter(!is.na(everCR_100)) %>% 
    filter(dr_car >=1, dr_car <11 ),
    'maaslin_cr100t_0_10',
    fixed_effects = c('everCR_100' ),
    random_effects = c('mrn'),
    standardize = FALSE, plot_heatmap = TRUE)
#

res_cr100_1_21<- Maaslin2(
    masdat, 
    sample_data(phy)%>% data.frame() %>% filter(!is.na(everCR_100)) %>% 
    filter(dr_car >=1, dr_car <22 ),
    'maaslin_cr100t_1_21',
    fixed_effects = c('everCR_100' ),
    random_effects = c('mrn'),
    standardize = FALSE, plot_heatmap = TRUE)
#



res_bestres_3lvl_100_1_10<- Maaslin2(
    masdat, 
    sample_data(phy)%>% data.frame() %>% filter(!is.na(bestres_3lvl_100)) %>% 
    filter(dr_car >=1, dr_car <11 ),
    'maaslin_bestres_3lvl_100t_1_10',
    fixed_effects = c('bestres_3lvl_100' ),
    random_effects = c('mrn'),
    reference = 'bestres_3lvl_100,CR',
    standardize = FALSE, plot_heatmap = TRUE)
#


# 
# 

# res_objres_100_1_10 <- not sig
res_objres_100_1_10<- Maaslin2(
    masdat,
    sample_data(phy)%>% data.frame() %>% filter(!is.na(objres_100)) %>%
    filter(dr_car >=1, dr_car <11 ),
    'maaslin_objres_100t_1_10',
    fixed_effects = c('objres_100' ),
    random_effects = c('mrn'),
    standardize = FALSE, plot_heatmap = TRUE)


# not sig
res_crs_tox_sig_pre_car_0<- Maaslin2(
    masdat,
    sample_data(phy)%>% data.frame() %>% filter(!is.na(tox_sig)) %>%
    filter(dr_car<=0 ),
    'maaslin_tox_sigt_pre_car',
    fixed_effects = c('tox_sig' ),
    random_effects = c('mrn'),
    standardize = FALSE, plot_heatmap = TRUE)


# not sig
res_crs_3_4_pre_car_0<- Maaslin2(
    masdat,
    sample_data(phy)%>% data.frame() %>% filter(!is.na(crs_3_4)) %>%
    filter(dr_car<=0 ),
    'maaslin_crs_3_4t_pre_car',
    fixed_effects = c('crs_3_4' ),
    random_effects = c('mrn'),
    standardize = FALSE, plot_heatmap = TRUE)


# not sig
res_icans_3_4_pre_car_0<- Maaslin2(
    masdat,
    sample_data(phy)%>% data.frame() %>% filter(!is.na(icans_3_4)) %>%
    filter(dr_car<=0 ),
    'maaslin_icans_3_4t_pre_car',
    fixed_effects = c('icans_3_4' ),
    random_effects = c('mrn'),
    standardize = FALSE, plot_heatmap = TRUE)

    
```

## Alpha diversity over time
```{r}
###

min_range = -7
max_range = 30


n_summary_samples <- plot_df %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  tally()

n_summary_patients<- plot_df %>% 
  filter(dr_car >min_range,  dr_car <max_range) %>% 
  group_by(record_id) %>% 
  slice(1) %>% 
  ungroup() %>% 
  tally()

p1 <- plot_df %>%
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  ggplot(aes(x=dr_car, y=InvSimpson)) +
  geom_jitter(size=3, alpha=.7, width = 0.5) +
  geom_smooth(alpha = 0.1)+
  scale_x_continuous(breaks = seq (-10,30, 5)) +
  labs(
 #   subtitle = "Day 100 response by inv. Simpson over time",
    caption = paste0("n_samples = ", n_summary_samples, "; n_patients = ", n_summary_patients, 
                     "\ntime range (days) [", min_range, " to ", max_range, "]" ),
    x = 'Days from CAR-T',
    y = 'Inverse Simpson') +
  guides(color = guide_legend( title = "",nrow = 1))+
  scale_color_manual(values = msk_palette("main")) +
  theme_msk() +
  theme (legend.position = "top")+ 
  set_msk_aes()

p1

########

outcome_criteria <- "everCR_100"

min_range = -7
max_range = 30


n_summary_samples <- plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  tally()

n_summary_patients<-  plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  group_by(record_id) %>% 
  slice(1) %>% 
  ungroup() %>% 
  tally()

p2 <- plot_df %>%
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  ggplot(aes(x=dr_car, y=InvSimpson, color = get(outcome_criteria) )) +
  geom_jitter(size=3, alpha=.7, width = 0.5) +
  geom_smooth(alpha = 0.1)+
  scale_x_continuous(breaks = seq (min_range ,max_range,  5)) +
  labs(
    #   subtitle = "Day 100 response by inv. Simpson over time",
    caption = paste0("n_samples = ", n_summary_samples, "; n_patients = ", n_summary_patients, 
                     "\ntime range (days) [", min_range, " to ", max_range, "]","\n" ),
    x = 'Days from CAR-T',
    y = 'Inverse Simpson') +
  guides(color = guide_legend( title = paste0(outcome_criteria, " "),nrow = 1))+
  scale_color_manual(values = msk_palette("main")) +
  theme_msk() +
  theme (legend.position = "top")+ 
  set_msk_aes()


########


########

outcome_criteria <- "objres_100"

min_range = -7
max_range = 30


n_summary_samples <- plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  tally()

n_summary_patients<-  plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  group_by(record_id) %>% 
  slice(1) %>% 
  ungroup() %>% 
  tally()

p3 <- plot_df %>%
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  ggplot(aes(x=dr_car, y=InvSimpson, color = get(outcome_criteria) )) +
  geom_jitter(size=3, alpha=.7, width = 0.5) +
  geom_smooth(alpha = 0.1)+
  scale_x_continuous(breaks = seq (min_range ,max_range,  5)) +
  labs(
    #   subtitle = "Day 100 response by inv. Simpson over time",
    caption = paste0("n_samples = ", n_summary_samples, "; n_patients = ", n_summary_patients, 
                     "\ntime range (days) [", min_range, " to ", max_range, "]", "\n" ),
    x = 'Days from CAR-T',
    y = 'Inverse Simpson') +
  guides(color = guide_legend( title = paste0(outcome_criteria, " "),nrow = 1))+
  scale_color_manual(values = msk_palette("main")) +
  theme_msk() +
  theme (legend.position = "top")+ 
  set_msk_aes()



########

outcome_criteria <- "crs.factor"

min_range = -7
max_range = 30


n_summary_samples <- plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  tally()

n_summary_patients<-  plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  group_by(record_id) %>% 
  slice(1) %>% 
  ungroup() %>% 
  tally()

p4 <- plot_df %>%
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  ggplot(aes(x=dr_car, y=InvSimpson, color = get(outcome_criteria) )) +
  geom_jitter(size=3, alpha=.7, width = 0.5) +
  geom_smooth(alpha = 0.1)+
  scale_x_continuous(breaks = seq (min_range ,max_range,  5)) +
  labs(
    #   subtitle = "Day 100 response by inv. Simpson over time",
    caption = paste0("n_samples = ", n_summary_samples, "; n_patients = ", n_summary_patients, 
                     "\ntime range (days) [", min_range, " to ", max_range, "]", "\n" ),
    x = 'Days from CAR-T',
    y = 'Inverse Simpson') +
  guides(color = guide_legend( title = paste0(outcome_criteria, " "),nrow = 1))+
  scale_color_manual(values = msk_palette("main")) +
  theme_msk() +
  theme (legend.position = "top")+ 
  set_msk_aes()





########

outcome_criteria <- "icans.factor"

min_range = -7
max_range = 30


n_summary_samples <- plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  tally()

n_summary_patients<-  plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  group_by(record_id) %>% 
  slice(1) %>% 
  ungroup() %>% 
  tally()

p5 <- plot_df %>%
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  ggplot(aes(x=dr_car, y=InvSimpson, color = get(outcome_criteria) )) +
  geom_jitter(size=3, alpha=.7, width = 0.5) +
  geom_smooth(alpha = 0.1)+
  scale_x_continuous(breaks = seq (min_range ,max_range,  5)) +
  labs(
    #   subtitle = "Day 100 response by inv. Simpson over time",
    caption = paste0("n_samples = ", n_summary_samples, "; n_patients = ", n_summary_patients, 
                     "\ntime range (days) [", min_range, " to ", max_range, "]", "\n" ),
    x = 'Days from CAR-T',
    y = 'Inverse Simpson') +
  guides(color = guide_legend( title = paste0(outcome_criteria, " "),nrow = 1))+
  scale_color_manual(values = msk_palette("main")) +
  theme_msk() +
  theme (legend.position = "top")+ 
  set_msk_aes()


########

outcome_criteria <- "car_t_product.factor"

min_range = -7
max_range = 30


n_summary_samples <- plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  tally()

n_summary_patients<-  plot_df %>% 
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  group_by(record_id) %>% 
  slice(1) %>% 
  ungroup() %>% 
  tally()
p6 <- plot_df %>%
  filter(!is.na(get(outcome_criteria))) %>% 
  filter(car_t_product.factor != "Lisocabtagene maraleucel") %>% 
  filter(dr_car > min_range,  dr_car < max_range) %>% 
  ggplot(aes(x=dr_car, y=InvSimpson, color = get(outcome_criteria) )) +
  geom_jitter(size=3, alpha=.7, width = 0.5) +
  geom_smooth(alpha = 0.1)+
  scale_x_continuous(breaks = seq (min_range ,max_range,  5)) +
  labs(
    #   subtitle = "Day 100 response by inv. Simpson over time",
    caption = paste0("n_samples = ", n_summary_samples, "; n_patients = ", n_summary_patients, 
                     "\ntime range (days) [", min_range, " to ", max_range, "]", "\n" ),
    x = 'Days from CAR-T',
    y = 'Inverse Simpson') +
  guides(color = guide_legend( title = paste0(outcome_criteria, " "),nrow = 1))+
  scale_color_manual(values = c("#83276B","#4C8B2B","#006098")) +
  theme_msk() +
  theme (legend.position = "top",
             legend.title = element_blank())+ 
  set_msk_aes()
p1
p2
p3
p4
p5
p6


```



## Box plot
```{r}
### Best CR d100
p <- plot_df %>% 
  mutate(everCR_100 = factor(everCR_100)) %>% 
 # arrange(record_id, dr_car) %>% 
  #group_by(record_id, range) %>%
 # slice(1) %>% 
#  ungroup() %>%
  filter(!is.na(everCR_100)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="everCR_100", y ="InvSimpson",
             color = "everCR_100",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
   scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$everCR_100)),
    caption = "All samples (not time restriction)",
    x =  paste(label(plot_df$everCR_100)),
    y = 'Inverse Simpson')

p


p <- plot_df %>% 
  mutate(everCR_100 = factor(everCR_100)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(!is.na(everCR_100)) %>%
   filter(!is.na(range)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="everCR_100", y ="InvSimpson",
             color = "everCR_100",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
   scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$everCR_100)),
    caption = "Earliest sample of each time window was selected",
    x =  paste(label(plot_df$everCR_100)),
    y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
   #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)

p <- plot_df %>% 
  mutate(everCR_100 = factor(everCR_100)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  summarise( InvSimpson = median (InvSimpson)) %>% 
  ungroup() %>%
  left_join(plot_df %>% 
              select(record_id, range, everCR_100), by = c("record_id", "range")) %>% 
  group_by(record_id, range ) %>% 
  slice(1) %>% 
  ungroup() %>% 
  filter(!is.na(everCR_100)) %>% 
     filter(!is.na(range)) %>% 
  mutate(everCR_100 = factor(everCR_100)) %>% 
  ggboxplot( x="everCR_100", y ="InvSimpson",
             color = "everCR_100",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
   scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$everCR_100)),
    caption = "Median inverse Simpson was calculated per patient per time window",
    x =  paste(label(plot_df$everCR_100)),
    y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
   #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)
       

### Best response 3 levels

my_comparisons <- list( c("CR", "PR"), c("CR", "SD/PD"), c("PR", "SD/PD") )

p <- plot_df %>% 
  mutate(bestres_3lvl_100 = factor(bestres_3lvl_100)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(!is.na(bestres_3lvl_100)) %>% 
       filter(!is.na(range)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="bestres_3lvl_100", y ="InvSimpson",
             color = "bestres_3lvl_100",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means(comparisons = my_comparisons)+
  #stat_compare_means(label.y = 43)   +  # Global comparison
  scale_color_manual(values = c("#007CBA", "#40B4E5", "#E92076","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$bestres_3lvl_100)),
       caption = "Earliest sample of each time window was selected",
       x =  paste(label(plot_df$bestres_3lvl_100)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
   #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)


p <- plot_df %>% 
  mutate(bestres_3lvl_100 = factor(bestres_3lvl_100)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  summarise( InvSimpson = median (InvSimpson)) %>% 
  ungroup() %>%
  left_join(plot_df %>% 
              select(record_id, range, bestres_3lvl_100), by = c("record_id", "range")) %>% 
  group_by(record_id, range ) %>% 
  slice(1) %>% 
  ungroup() %>% 
  filter(!is.na(bestres_3lvl_100)) %>% 
         filter(!is.na(range)) %>% 
  mutate(bestres_3lvl_100 = factor(bestres_3lvl_100)) %>% 
  ggboxplot( x="bestres_3lvl_100", y ="InvSimpson",
             color = "bestres_3lvl_100",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means(comparisons = my_comparisons)+
  #stat_compare_means(label.y = 43)   +
  scale_color_manual(values = c("#007CBA", "#40B4E5", "#E92076","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$bestres_3lvl_100)),
       caption = "Median inverse Simpson was calculated per patient per time window",
       x =  paste(label(plot_df$bestres_3lvl_100)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)


### 
### Objective response
p <- plot_df %>% 
  mutate(objres_100 = factor(objres_100)) %>% 
  mutate (objres_100 = fct_rev(objres_100)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(!is.na(objres_100)) %>% 
   filter(!is.na(range)) %>% 
    #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="objres_100", y ="InvSimpson",
             color = "objres_100",
             add = "jitter") +
  
  stat_compare_means() +
  scale_color_manual(values = c("#83276B", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$objres_100)),
       caption = "Earliest sample of each time window was selected",
       x =  paste(label(plot_df$objres_100)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)

p <- plot_df %>% 
  mutate(objres_100 = factor(objres_100)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  summarise( InvSimpson = median (InvSimpson)) %>% 
  ungroup() %>%
  left_join(plot_df %>% 
              select(record_id, range, objres_100), by = c("record_id", "range")) %>% 
  group_by(record_id, range ) %>% 
  slice(1) %>% 
  ungroup() %>% 
  filter(!is.na(objres_100)) %>% 
  filter(!is.na(range)) %>% 
  mutate(objres_100 = fct_rev(factor(objres_100))) %>% 
  ggboxplot( x="objres_100", y ="InvSimpson",
             color = "objres_100",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#83276B", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$objres_100)),
       caption = "Median inverse Simpson was calculated per patient per time window",
       x =  paste(label(plot_df$objres_100)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)


##
### PFS 6
p <- plot_df %>% 
  mutate(pfs_6_ev_binary= factor(pfs_6_ev_binary)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(!is.na(pfs_6_ev_binary)) %>% 
  filter(!is.na(range)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="pfs_6_ev_binary", y ="InvSimpson",
             color = "pfs_6_ev_binary",
             add = "jitter") +
  
  stat_compare_means() +
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$pfs_6_ev_binary)),
       caption = "Earliest sample of each time window was selected",
       x =  paste(label(plot_df$pfs_6_ev_binary)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)

p <- plot_df %>% 
  mutate(pfs_6_ev_binary= factor(pfs_6_ev_binary)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  summarise( InvSimpson = median (InvSimpson)) %>% 
  ungroup() %>%
  left_join(plot_df %>% 
              select(record_id, range, pfs_6_ev_binary), by = c("record_id", "range")) %>% 
  group_by(record_id, range ) %>% 
  slice(1) %>% 
  ungroup() %>% 
  filter(!is.na(pfs_6_ev_binary)) %>%
  filter(!is.na(range)) %>% 
  mutate(pfs_6_ev_binary= fct_rev(factor(pfs_6_ev_binary))) %>% 
  ggboxplot( x="pfs_6_ev_binary", y ="InvSimpson",
             color = "pfs_6_ev_binary",
             add = "jitter") +
  
  stat_compare_means() +
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$pfs_6_ev_binary)),
       caption = "Median inverse Simpson was calculated per patient per time window",
       x =  paste(label(plot_df$pfs_6_ev_binary)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)

###
### Best CRS -2 


p <- plot_df %>% 
  mutate(crs_2_4 = factor(crs_2_4)) %>% 
  # arrange(record_id, dr_car) %>% 
  # group_by(record_id, range) %>%
  # slice(1) %>% 
  ungroup() %>%
  filter(!is.na(crs_2_4)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="crs_2_4", y ="InvSimpson",
             color = "crs_2_4",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$crs_2_4)),
       caption = "All samples",
       x =  paste(label(plot_df$crs_2_4)),
       y = 'Inverse Simpson')

p







p <- plot_df %>% 
  mutate(crs_3_4 = factor(crs_3_4)) %>% 
  # arrange(record_id, dr_car) %>% 
  # group_by(record_id, range) %>%
  # slice(1) %>% 
  ungroup() %>%
  filter(!is.na(crs_3_4)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="crs_3_4", y ="InvSimpson",
             color = "crs_3_4",
             add = "jitter") +
  
  stat_compare_means() +
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$crs_3_4)),
       caption = "All samples",
       x =  paste(label(plot_df$crs_3_4)),
       y = 'Inverse Simpson')

p




p <- plot_df %>% 
  mutate(crs_2_4 = factor(crs_2_4)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(!is.na(crs_2_4)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="crs_2_4", y ="InvSimpson",
             color = "crs_2_4",
             add = "jitter") +
  
  stat_compare_means() +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$crs_2_4)),
       caption = "Earliest sample of each time window was selected",
       x =  paste(label(plot_df$crs_2_4)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
   #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)


p <- plot_df %>% 
  mutate(crs_2_4 = factor(crs_2_4)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  summarise( InvSimpson = median (InvSimpson)) %>% 
  ungroup() %>%
  left_join(plot_df %>% 
              select(record_id, range, crs_2_4), by = c("record_id", "range")) %>% 
  group_by(record_id, range ) %>% 
  slice(1) %>% 
  ungroup() %>% 
  filter(!is.na(crs_2_4)) %>% 
  mutate(crs_2_4 = factor(crs_2_4)) %>% 
  ggboxplot( x="crs_2_4", y ="InvSimpson",
             color = "crs_2_4",
             add = "jitter") +
   scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$crs_2_4)),
       caption = "Median inverse Simpson was calculated per patient per time window",
       x =  paste(label(plot_df$crs_2_4)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
   #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)

###



p <- plot_df %>% 
  mutate(icans_2_4 = factor(icans_2_4)) %>% 
  # arrange(record_id, dr_car) %>% 
  # group_by(record_id, range) %>%
  # slice(1) %>% 
  ungroup() %>%
  filter(!is.na(icans_2_4)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="icans_2_4", y ="InvSimpson",
             color = "icans_2_4",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$icans_2_4)),
       caption = "All samples",
       x =  paste(label(plot_df$icans_2_4)),
       y = 'Inverse Simpson')

p


p <- plot_df %>% 
  mutate(icans_2_4 = factor(icans_2_4)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(!is.na(icans_2_4)) %>% 
  #  filter(range %in% range_criteria) %>% 
  ggboxplot( x="icans_2_4", y ="InvSimpson",
             color = "icans_2_4",
             add = "jitter") +
  
  stat_compare_means() +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$icans_2_4)),
       caption = "Earliest sample of each time window was selected",
       x =  paste(label(plot_df$icans_2_4)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)


p <- plot_df %>% 
  mutate(icans_2_4 = factor(icans_2_4)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  summarise( InvSimpson = median (InvSimpson)) %>% 
  ungroup() %>%
  left_join(plot_df %>% 
              select(record_id, range, icans_2_4), by = c("record_id", "range")) %>% 
  group_by(record_id, range ) %>% 
  slice(1) %>% 
  ungroup() %>% 
  filter(!is.na(icans_2_4)) %>% 
  mutate(icans_2_4 = factor(icans_2_4)) %>% 
  ggboxplot( x="icans_2_4", y ="InvSimpson",
             color = "icans_2_4",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#007CBA", "#E77441", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$icans_2_4)),
       caption = "Median inverse Simpson was calculated per patient per time window",
       x =  paste(label(plot_df$icans_2_4)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)



p <- plot_df %>% 
  mutate(crs_max.factor = factor(crs_max.factor)) %>% 
  # arrange(record_id, dr_car) %>% 
  # group_by(record_id, range) %>%
  # summarise( InvSimpson = median (InvSimpson)) %>% 
  # ungroup() %>%
  # left_join(plot_df %>% 
  #             select(record_id, range, crs_max.factor), by = c("record_id", "range")) %>% 
  # group_by(record_id, range ) %>% 
  # slice(1) %>% 
  # ungroup() %>% 
  filter(!is.na(crs_max.factor)) %>% 
  mutate(crs_max.factor = factor(crs_max.factor)) %>% 
  ggboxplot( x="crs_max.factor", y ="InvSimpson",
             color = "crs_max.factor",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#B6B6B1", "#8FC7E8", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$crs_max.factor)),
       caption = "All samples",
       x =  paste(label(plot_df$crs_max.factor)),
       y = 'Inverse Simpson')


plot_df %>% 
  mutate(icans_max.factor = factor(icans_max.factor)) %>% 
  # arrange(record_id, dr_car) %>% 
  # group_by(record_id, range) %>%
  # summarise( InvSimpson = median (InvSimpson)) %>% 
  # ungroup() %>%
  # left_join(plot_df %>% 
  #             select(record_id, range, icans_max.factor), by = c("record_id", "range")) %>% 
  # group_by(record_id, range ) %>% 
  # slice(1) %>% 
  # ungroup() %>% 
  filter(!is.na(icans_max.factor)) %>% 
  mutate(icans_max.factor = factor(icans_max.factor)) %>% 
  ggboxplot( x="icans_max.factor", y ="InvSimpson",
             color = "icans_max.factor",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#B6B6B1", "#8FC7E8", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$icans_max.factor)),
       caption = "All samples",
       x =  paste(label(plot_df$icans_max.factor)),
       y = 'Inverse Simpson')


p <- plot_df %>% 
  mutate(crs_max.factor = factor(crs_max.factor)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  summarise( InvSimpson = median (InvSimpson)) %>% 
  ungroup() %>%
  left_join(plot_df %>% 
              select(record_id, range, crs_max.factor), by = c("record_id", "range")) %>% 
  group_by(record_id, range ) %>% 
  slice(1) %>% 
  ungroup() %>% 
  filter(!is.na(crs_max.factor)) %>% 
  mutate(crs_max.factor = factor(crs_max.factor)) %>% 
  ggboxplot( x="crs_max.factor", y ="InvSimpson",
             color = "crs_max.factor",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#B6B6B1", "#8FC7E8", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$crs_max.factor)),
       caption = "Median inverse Simpson was calculated per patient per time window",
       x =  paste(label(plot_df$crs_max.factor)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)


p <- plot_df %>% 
  mutate(icans_max.factor = factor(icans_max.factor)) %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  summarise( InvSimpson = median (InvSimpson)) %>% 
  ungroup() %>%
  left_join(plot_df %>% 
              select(record_id, range, icans_max.factor), by = c("record_id", "range")) %>% 
  group_by(record_id, range ) %>% 
  slice(1) %>% 
  ungroup() %>% 
  filter(!is.na(icans_max.factor)) %>% 
  mutate(icans_max.factor = factor(icans_max.factor)) %>% 
  ggboxplot( x="icans_max.factor", y ="InvSimpson",
             color = "icans_max.factor",
             add = "jitter") +
  scale_y_continuous(expand = (add = c(0.2, 0)))+
  stat_compare_means() +
  scale_color_manual(values = c("#B6B6B1", "#8FC7E8", "#007CBA","#F29934","#E92076")) +
  theme_msk() +
  set_msk_aes() +
  labs(subtitle = paste(label(plot_df$icans_max.factor)),
       caption = "Median inverse Simpson was calculated per patient per time window",
       x =  paste(label(plot_df$icans_max.factor)),
       y = 'Inverse Simpson')

facet(p,
      facet.by = "range",
      short.panel.labs =F,
      #   panel.labs.background	= list(fill = "#FAFAFA"),
      panel.labs.font.x = list (size =10,   face = "bold")
)

```
## Survival analysis by alpha diversity
```{r}

time_range = "pre-car"
# define above and blow median a-diversity for each time frame
div_data<- plot_df %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(range == time_range) %>% 
  filter( !is.na(InvSimpson)) %>% 
  mutate(med_div = ifelse (InvSimpson > median(InvSimpson),"> med div", "<=med div"),
         med_fiv = factor( med_div, levels = c("High div.", "Low div."))) 
  
         

pfs = survfit(Surv(tt_pfs_d/30.42, ev_pfs)~med_div, data = div_data)


pfs_plot <-ggsurvplot(pfs, data =div_data ,
                      risk.table = TRUE, tables.height = 0.15,
                      tables.theme = theme_cleantable(),
                      palette = rev(c("#007CBA", "#EFB3CB")),
                      censor=T,
                      conf.int = F,
                      legend = c("none"),
                      break.x.by=3,
                      xlim = c(0,13),
                      axes.offset=FALSE,
                      xlab="Time in Months",
                      ylab ="PFS",
                      pval = T,
                      ggtheme = theme_msk()) 

pfs_plot$plot <- pfs_plot$plot + labs(
  caption  = paste("Sample time window: ", time_range, "\nEarliest sample selected"))
pfs_plot$table  <-  pfs_plot$table   +
  theme(plot.title = element_text(size = 9, face = "italic"))

pfs_plot

tbl_survfit(
  pfs,
  times = c(12, 24),
  label = "{time} Months"
) %>% 
  modify_caption("PFS Rates")


tbl_survfit(
  pfs,
  probs = c(0.5),
  header_estimate = "**Months**"
) %>% 
  modify_caption("Median PFS")        

########## new time range         

time_range = "1-10"
div_data<- plot_df %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(range == time_range) %>% 
  filter( !is.na(InvSimpson)) %>% 
  mutate(med_div = ifelse (InvSimpson > median(InvSimpson),"> med div", "<=med div"),
         med_fiv = factor( med_div, levels = c("High div.", "Low div."))) 
  
# landmarking         
df_10_pfs = div_data %>% 
  mutate(tt_pfs_d = tt_pfs_d - 10) %>% 
  filter(tt_pfs_d >=0)


pfs = survfit(Surv(tt_pfs_d/30.42, ev_pfs)~med_div, data = df_10_pfs)


pfs_plot <-ggsurvplot(pfs, data =df_10_pfs ,
                      risk.table = TRUE, tables.height = 0.15,
                      tables.theme = theme_cleantable(),
                      palette = rev(c("#007CBA", "#EFB3CB")),
                      censor=T,
                      conf.int = F,
                      legend = c("none"),
                      break.x.by=3,
                      xlim = c(0,13),
                      axes.offset=FALSE,
                      xlab="Time in Months",
                      ylab ="PFS",
                      pval = T,
                      ggtheme = theme_msk()) 

pfs_plot$plot <- pfs_plot$plot + labs(
  caption  = paste("Sample time window: ", time_range, "\nEarliest sample selected\nlandmark day 10"))
pfs_plot$table  <-  pfs_plot$table   +
  theme(plot.title = element_text(size = 9, face = "italic"))


#pdf("pfs_entire_cohort.pdf")
pfs_plot
#dev.off()

tbl_survfit(
  pfs,
  times = c(12, 24),
  label = "{time} Months"
) %>% 
  modify_caption("PFS Rates")


tbl_survfit(
  pfs,
  probs = c(0.5),
  header_estimate = "**Months**"
) %>% 
  modify_caption("Median PFS")      



time_range = "pre-car"
div_data<- plot_df %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(range == time_range) %>% 
  filter( !is.na(InvSimpson)) %>% 
  mutate(med_div = ifelse (InvSimpson > median(InvSimpson),"> med div", "<=med div"),
         med_fiv = factor( med_div, levels = c("High div.", "Low div."))) 



os = survfit(Surv(tt_os_d/30.42, ev_os)~med_div, data = div_data)


os_plot <-ggsurvplot(os, data =div_data ,
                      risk.table = TRUE, tables.height = 0.15,
                      tables.theme = theme_cleantable(),
                      palette = rev(c("#007CBA", "#EFB3CB")),
                      censor=T,
                      conf.int = F,
                      legend = c("none"),
                      break.x.by=3,
                      xlim = c(0,13),
                      axes.offset=FALSE,
                      xlab="Time in Months",
                      ylab ="os",
                      pval = T,
                      ggtheme = theme_msk()) 


os_plot$plot <- os_plot$plot + labs(
  caption  = paste("Sample time window: ", time_range,"\nEarliest sample selected"))
os_plot$table  <-  os_plot$table   +
  theme(plot.title = element_text(size = 9, face = "italic"))


#pdf("os_entire_cohort.pdf")
os_plot
#dev.off()

tbl_survfit(
  os,
  times = c(12, 24),
  label = "{time} Months"
) %>% 
  modify_caption("os Rates")


tbl_survfit(
  os,
  probs = c(0.5),
  header_estimate = "**Months**"
) %>% 
  modify_caption("Median os")        
########## new time range         

time_range = "1-10"
div_data<- plot_df %>% 
  arrange(record_id, dr_car) %>% 
  group_by(record_id, range) %>%
  slice(1) %>% 
  ungroup() %>%
  filter(range == time_range) %>% 
  filter( !is.na(InvSimpson)) %>% 
  mutate(med_div = ifelse (InvSimpson > median(InvSimpson),"> med div", "<=med div"),
         med_fiv = factor( med_div, levels = c("High div.", "Low div."))) 


df_10_os = div_data %>% 
  mutate(tt_os_d = tt_os_d - 10) %>% 
  filter(tt_os_d >=0)


os = survfit(Surv(tt_os_d/30.42, ev_os)~med_div, data = df_10_os)


os_plot <-ggsurvplot(os, data =df_10_os ,
                      risk.table = TRUE, tables.height = 0.15,
                      tables.theme = theme_cleantable(),
                      palette = rev(c("#007CBA", "#EFB3CB")),
                      censor=T,
                      conf.int = F,
                      legend = c("none"),
                      break.x.by=3,
                      xlim = c(0,13),
                      axes.offset=FALSE,
                      xlab="Time in Months",
                      ylab ="os",
                      pval = T,
                      ggtheme = theme_msk()) 

os_plot$plot <- os_plot$plot + labs(
  caption  = paste("Sample time window: ", time_range, "\nEarliest sample selected\nlandmark day 10"))
os_plot$table  <-  os_plot$table   +
  theme(plot.title = element_text(size = 9, face = "italic"))


#pdf("os_entire_cohort.pdf")
os_plot
#dev.off()

tbl_survfit(
  os,
  times = c(12, 24),
  label = "{time} Months"
) %>% 
  modify_caption("os Rates")


tbl_survfit(
  os,
  probs = c(0.5),
  header_estimate = "**Months**"
) %>% 
  modify_caption("Median os")      


    
```
