---
title: "The stool samples for the 67 pateints"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Getting the stool samples path for the 67 patients in the diet cohort right now


```{r}
library(tidyverse)
```

```{r}
source('~/db_connect_simple.R')
connect_database(config_file = "~/dbConfig.txt")

get_table_from_database_predefined_filter('asv_alpha_diversity_ag')

allo <- get_table_from_database('patient_allo_ag')
pheno <- read_csv('../data/pheno/final_pheno.csv')
castori <- get_table_from_database('samples_castori_ag')
```


```{r}
samples <- pheno %>% 
  left_join(allo %>% 
              filter(str_detect(indication, fixed('initial', ignore_case = T))) %>% 
              select(mrn, hct)) %>% 
  left_join(castori %>% 
              filter(sampletype == 'Stool' | sampletype == 'stool') %>% 
              select(mrn, sampleid, datecollection), by = 'mrn') %>% 
  mutate(drt = datecollection - hct) %>% 
  inner_join(asv_alpha_diversity_ag %>% 
               select(sampleid, simpson_reciprocal:path_pool), by = 'sampleid')

# 50 we don't have counts data   

samples %>% 
  write_csv('../data/pheno/samples_pheno_67.csv')

samples <- read_csv('../data/pheno/samples_pheno_67.csv')

samples %>% 
  write_tsv('../data/pheno/samples_pheno_67.tsv')
```


```{r}
# got the goddamn path of the files
paths <- read_delim('../../upload_sample_NCBI/data/demultiplexed_path/list_demultiplexed_samples_castori.txt', delim = ' ', col_names = c('folder','pool','sampleid'))  %>% 
  select(-pool) %>% 
  mutate(direction = if_else(str_detect(folder, 'R1.fastq.gz'), 'R1', 'R2')) %>% 
  filter(sampleid %in% samples$sampleid) %>% 
  spread('direction', 'folder')

# so there are some samples in our cohort currently not in the above paths
missing <- setdiff(samples$sampleid, paths$sampleid)

other <- read_delim('../../upload_sample_NCBI/data/demultiplexed_path/list_demultiplexed_samples_extra.txt', delim = ' ', col_names = c('folder','pool','sampleid'))  %>% 
  select(-pool) %>% 
  mutate(direction = if_else(str_detect(folder, 'R1.fastq.gz'), 'R1', 'R2')) %>% 
  filter(sampleid %in% missing) %>% 
  spread('direction', 'folder')

other_sampleid <- other$sampleid
```


```{r}
# another 10 missing ....
missing10 <- setdiff(missing, other$sampleid)
 

last <- asv_alpha_diversity_ag %>% 
  filter(sampleid %in% missing10) %>% 
  mutate(R1 = str_glue('{path_pool}/{oligos_id}_R1.fastq.gz'),
         R2 = str_replace(R1, '_R1.fastq.gz','_R2.fastq.gz')) %>% 
  select(sampleid, R1, R2)

other_alpha <- asv_alpha_diversity_ag %>% 
  filter(sampleid %in% other_sampleid) %>% 
  mutate(R1 = str_glue('{path_pool}/{oligos_id}_R1.fastq.gz'),
         R2 = str_replace(R1, '_R1.fastq.gz','_R2.fastq.gz')) %>% 
  select(sampleid, R1, R2)

other_alpha$path_pool
```


```{r}
all_path <- bind_rows(
  paths, other_alpha, last
) %>% 
  rename(`sample-id` = sampleid, 
         `forward-absolute-filepath` = R1,
         `reverse-absolute-filepath` = R2)

all_path %>% 
  write_tsv('../data/qiime/manifest.tsv')
```

**Summary: just use the alpha diversity table, and glue the path_pool with oligos_id and then you got the full file path!**

```{r}
# the pheno info for these samples
meta <- samples %>% 
  select(sampleid, drt, intensity, source, simpson_reciprocal, shannon, count_total)


meta %>% 
  write_tsv('../data/qiime/diet_67_meta.tsv')


stool_meta <- read_tsv('../data/qiime/diet_67_meta.tsv')
```

```{r}
# look at drt in these 1002 samples
range(samples$drt)
```



# upload the tables to database 

```{r}
bile <- read_csv('../../picrust2/data/normalized_picrust2_pred_metagenome_KO_secondary_bile_counts.csv')

source('/Volumes/vandenBrinkLab/gomesa/2Angel/scripts/create_table_AD.R')

create_table_prepare(bile)

create_table_type(2)

# upload data

source('/Volumes/vandenBrinkLab/gomesa/2Angel/scripts/upload_data_from_query_OTU_AD.R')

upload_data_from_query_OTU(2)

get_table_from_database('bile_picrust_ad')
```


