---
title: "The mouse data Enterococcus"
author: "Angel"
date: "2022-07-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(vdbR)
library(readxl)
library(ggpubr)
connect_database()
get_table_from_database('asv_alpha_diversity_ag')
get_table_from_database('asv_annotation_blast_ag')
get_table_from_database('asv_annotation_blast_color_ag')
```


See the relab and absolute copy numbers of these data 

```{r}
q <- read_excel('../data/Tsoni_59samples_pool1154_16SqPCR_Results_07112022.xlsx')
updated_q <- q %>%
  rename(sampleid=Sample.ID,
         copies_16s_per_g_pre = "16S COPIES/G") %>% 
  mutate(
    g = `16S COPY NUMBER`/ copies_16s_per_g_pre,
    copies_16s_bounded = ifelse(
      `16S COPY NUMBER` < 12800, 
      (12800*.5),
      ifelse(
        `16S COPY NUMBER` > 2e8,
        1.5 * (2e8),
        `16S COPY NUMBER`
      )
    )
  ) %>% 
  mutate(copies_16s_per_g =  copies_16s_bounded/g)  # use the *copies_16s_per_g*
```


```{r}
pheno <- read_csv('../data/correct_flourfenicol_biapenem_cfucounts.csv') %>% 
  filter(!is.na(Log_CFUs_per_GramStool)) # there is one mouse that no stool was collected
# Only C014165 has no stool pellet so it wasn’t given to sequencing.
# C014184: That tube has stool. It just didn’t grow on our plate selective for enterococcus.(no colony count)

sampleids <- asv_alpha_diversity_ag %>% 
  filter(str_detect(oligos_id, 'pool1154'))
cts <- get_counts_subset(sampleids$sampleid)
```

```{r cts_entero}
cts_entero <- cts %>% 
  left_join(asv_annotation_blast_ag %>% 
              select(asv_key, genus)) %>% 
  group_by(sampleid, genus) %>% 
  summarise(relab = sum(count_relative)) %>% 
  filter(genus == 'Enterococcus') %>% 
  full_join(updated_q %>% 
               select(sampleid  = `Tube ID`, copies_16s_per_g )) %>% 
  mutate(entero_copy = copies_16s_per_g*relab)  %>% 
  inner_join(pheno %>% 
               select(sampleid = TubeID, Treatment:Mouse_identifier, Log_CFUs_per_GramStool )) %>% 
  inner_join(sampleids %>% 
              select(sampleid:simpson_reciprocal)) %>% 
  relocate(relab, .after = simpson_reciprocal) %>% 
  relocate(entero_copy, .after = relab)  %>% 
  mutate(abx = str_extract(Treatment, '^.+\\+'),
         abx = str_replace(abx, ' \\+', ''),
        grouping = str_extract(Treatment, '\\+.+$'),
         grouping = str_replace(grouping, '\\+ ', ''),
         grouping = factor(grouping, levels = c('Plain Hydrogel','Sucrose','Naked Smoothie','Canned Peaches')) ,
         sid = str_glue('{grouping}__Day{Day}__{Mouse_identifier}'),
         faceting_var = str_glue('{grouping}__Day{Day}')) %>% 
  relocate(Log_CFUs_per_GramStool, .before = 'grouping') %>% 
  relocate(abx, after = 'grouping') %>% 
  gather('key', 'value', relab:Log_CFUs_per_GramStool) %>% 
  mutate(key = factor(key, levels = c('Log_CFUs_per_GramStool','entero_copy','relab')))

cts_entero %>% 
  write_csv('../data/121_mouse_exp.csv')

# to only plotting the Biapenem for now 
bia_df <- cts_entero %>% 
  filter(abx == 'Biapenem')
```
```{r}
rows <- bia_df %>% 
  ggscatter(x = 'sampleid', y = 'value', title = 'Biapenem',
            ylab = 'Enterococcus') +
  facet_grid(key~ faceting_var, scales = 'free') +
  theme_linedraw() +
  theme(legend.position = 'none',
        strip.background = element_blank(),
        axis.text.x = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.title.x = element_blank())

ggsave('../data/121_mouse_exp.pdf', width = 15)
```


# bar plot and pcoa


```{r}
# also make bar graph and also pcoa
genus <- cts %>% 
  left_join(asv_annotation_blast_ag %>% select(asv_key, genus)) %>% 
  group_by(sampleid, genus) %>% 
  summarise(relab = sum(count_relative)) %>% 
  ungroup() %>% 
  left_join(pheno %>% select(sampleid = TubeID, Treatment:Mouse_identifier)) %>% 
  arrange(Treatment, Day, Mouse_identifier) %>% 
  mutate(sid = str_glue('{Treatment}__Day:{Day}__Mouse:{Mouse_identifier}'))


```


```{r}
# # find the genera > 0.1% in more than 10% samples
#  22/108 left 
keepgenera <- genus %>% 
  filter(relab > 0.01) %>% 
  count(genus) %>% 
  filter(n > floor(nrow(pheno) * 0.1)) %>% 
  pull(genus)

genus %>% ungroup %>% distinct(genus)
```




```{r}
# to find colors for these features
library(randomcoloR)
n <- length(keepgenera)
palette <- distinctColorPalette(n)
pie(rep(1, n), col=palette)

bar_palette <- genus %>%
  filter(genus %in% keepgenera) %>%
  distinct(genus) %>%
  arrange(genus) %>%
  mutate(color = palette)

bar_plot_pal <- deframe(bar_palette)


genus %>% 
  filter(genus %in% keepgenera) %>% 
  ggbarplot(x = 'sid', y = 'relab', fill = 'genus', color = 'white' , xlab = '', ylab = 'Relative abundance') +
  facet_grid(Day ~ Treatment, scales = 'free') +
  scale_fill_manual(values = bar_plot_pal) +
  theme( axis.text.x = element_text(angle=90, hjust=0, size = 6.5),
         legend.position = 'bottom') 

ggsave('../data/121_bar_mouse.pdf', height  = 14, width = 14)
```

```{r}
library(vegan)
# pcoa
keep_asv <- cts %>% 
  filter(count_relative > 0.0001) %>% 
  count(asv_key) %>% 
  filter(n > floor(nrow(pheno) * 0.1)) %>% 
  pull(asv_key)


# calculate the Bray curtis beta diversity 
cts_fil <- cts %>% 
  filter(asv_key %in% keep_asv) %>% 
  select(sampleid, asv_key,count_relative ) %>% 
  spread(key = 'asv_key', value = 'count_relative', fill = 0) %>% 
  column_to_rownames('sampleid')

dist_ <- vegdist(cts_fil, method = 'bray')
eigen <- cmdscale(dist_, eig = T)$eig
percent_var <- signif(eigen/sum(eigen), 3)*100
bc <- cmdscale(dist_, k = 2)
beta_df <- bc %>%
  as.data.frame() %>%
  rownames_to_column('sampleid')  %>% 
  full_join(pheno %>% select(sampleid = TubeID, Treatment:Mouse_identifier)) %>% 
  separate(Treatment, into = c('abx', 'treatment'), sep = ' \\+ ', remove = F) %>% 
  mutate(Day = factor(Day, levels = c(0, 3, 7)))
```


```{r}
pcoa_all <- beta_df %>% 
  ggscatter(x = 'V1', y = 'V2', color =  'Day', palette = 'lancet',ellipse = TRUE,
            title = 'color by Day',
            #facet.by = 'Day',
           # size = 'Day', shape = 'fiber',
            alpha = 0.5) +
  #facet_grid(Day ~ .) +
  labs(title = '') +
  xlab(paste0("PC 1 [",percent_var[1],"%]")) +
  ylab(paste0("PC 2 [",percent_var[2],"%]")) +
  theme(aspect.ratio=1, legend.position = 'bottom')
pcoa_all

ggsave('../data/121_pcoa_mouse.pdf', height  = 14, width = 14)
```

```{r}
beta_df %>% 
  ggscatter(x = 'V1', y = 'V2', color =  'abx',palette = 'lancet',ellipse = TRUE,
            title = 'color by abx',
            #facet.by = 'Day',
           # size = 'Day', shape = 'fiber',
            alpha = 0.5) +
  #facet_grid(Day ~ abx) +
  labs(title = '') +
  xlab(paste0("PC 1 [",percent_var[1],"%]")) +
  ylab(paste0("PC 2 [",percent_var[2],"%]")) +
  theme(aspect.ratio=1, legend.position = 'bottom')
```
```{r}
beta_df %>% 
  ggscatter(x = 'V1', y = 'V2', color =  'treatment',palette = 'lancet',ellipse = TRUE,
            title = 'color by treatment',
            #facet.by = 'Day',
           # size = 'Day', shape = 'fiber',
            alpha = 0.5) +
  #facet_grid(Day ~ abx) +
  labs(title = '') +
  xlab(paste0("PC 1 [",percent_var[1],"%]")) +
  ylab(paste0("PC 2 [",percent_var[2],"%]")) +
  theme(aspect.ratio=1, legend.position = 'bottom')
```
```{r}
beta_df %>% 
  ggscatter(x = 'V1', y = 'V2', color =  'Mouse_identifier',palette = 'lancet',ellipse = TRUE,
            title = 'color by Mouse_identifier',
            #facet.by = 'Day',
           # size = 'Day', shape = 'fiber',
            alpha = 0.5) +
  #facet_grid(Day ~ abx) +
  labs(title = '') +
  xlab(paste0("PC 1 [",percent_var[1],"%]")) +
  ylab(paste0("PC 2 [",percent_var[2],"%]")) +
  theme(aspect.ratio=1, legend.position = 'bottom')
```

# bar chart using the human palette

```{r}
dat <- cts %>% 
  select(asv_key:count_total, count_relative) %>% 
  left_join(asv_annotation_blast_color_ag %>% 
              select(asv_key,color_label_group_distinct), by = "asv_key")
```

```{r}
# there are some ASVs that don't have a color with it, but can use the color for the genus level
color_group <- dat %>% 
  split(is.na(.$color_label_group_distinct))
# find the genus for these asv

no_color <- color_group %>% 
  pluck('TRUE') %>% 
  distinct(asv_key) %>% 
  inner_join(asv_annotation_blast_ag %>% 
               select(asv_key, genus)) 
# find the colors for these genera
genera_colors <- no_color %>% 
  distinct(genus) %>% 
  inner_join(asv_annotation_blast_color_ag %>% 
               distinct(genus, color_label_group_distinct))
# the full df for the no color genera
no_color_df <- no_color %>% 
  left_join(genera_colors)
no_color_df_full <- color_group %>% 
  pluck('TRUE') %>% 
  select(-color_label_group_distinct) %>% 
  left_join(no_color_df %>% 
              select(- genus))
  
# so if the genus is unknown then it's gonna be assigned "other" gray color  
# the question is do we go one taxa level higher or make a new color base and shades for the new asv
# after discussing with Tsoni, we decided that it's ok to assign gray to the unknown genus 
# merge the new no_color_df_full to the original df
dat <- bind_rows(
  no_color_df_full,
  color_group %>% 
    pluck('FALSE')
) %>% left_join(pheno %>% 
               select(sampleid = TubeID, Treatment:Mouse_identifier, Log_CFUs_per_GramStool )) %>% 
  mutate(abx = str_extract(Treatment, '^.+\\+'),
         abx = str_replace(abx, ' \\+', ''),
        grouping = str_extract(Treatment, '\\+.+$'),
         grouping = str_replace(grouping, '\\+ ', ''),
         grouping = factor(grouping, levels = c('Plain Hydrogel','Sucrose','Naked Smoothie','Canned Peaches')) ,
         sid = str_glue('{grouping}__Day{Day}__{Mouse_identifier}'),
         faceting_var = str_glue('{grouping}__Day{Day}'))

dat_bia <- dat %>% 
  filter(abx == 'Biapenem')
```


```{r}
asv_color_set <- asv_annotation_blast_color_ag %>% 
  distinct(color,color_label_group_distinct,color_label_group,color_base) %>% 
  select(color_label_group_distinct, color) %>% 
  deframe()

cbd <- compute_beta_diversity_and_tsne(sampleid = dat_bia$sampleid, 
                                      taxonomy = dat_bia$color_label_group_distinct,
                                      count = dat_bia$count);
#compute beta diversity
cbd$compute_beta_diversity()


stacked_bar <- ggplot(dat_bia,aes(sid, count_relative, fill = color_label_group_distinct) ) +
  geom_bar(stat = "identity", position="fill", width = 1) +
  theme_classic() +
  facet_grid(. ~ faceting_var, scales = 'free') +
  theme_linedraw() +
  labs(title = '',
       y = 'Relative abundance') +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_blank(),
        axis.title.x = element_blank(),
        strip.background = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = asv_color_set)

stacked_bar
```


```{r}
library(cowplot)
p <- '../../miscellaneous/data//composition_legend.png'
legend <- ggdraw() +
  draw_image(magick::image_read(p),   scale = 0.9) 
legend 


top <- plot_grid(stacked_bar, legend, labels = 'none',rel_widths  = c(2, 0.5) )

ggsave('../data/121_bar_human_palette.pdf',
        width = 15, plot = stacked_bar)
```


```{r}
library(cowplot)
both <- plot_grid(rows, stacked_bar, nrow = 2,align = 'hv', axis = 'lr')
ggsave('../data/121_overlay.pdf',
        width = 10, plot = both)
```